--- binutils-2.17.50.0.9/configure   2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/configure-patched   2007-11-28 22:19:29.000000000 +0100
@@ -3799,7 +3799,7 @@
     # For an installed makeinfo, we require it to be from texinfo 4.4 or
     # higher, else we use the "missing" dummy.
     if ${MAKEINFO} --version \
-       | egrep 'texinfo[^0-9]*([1-3][0-9]|4\.[4-9]|[5-9])' >/dev/null 2>&1; then
+       | egrep 'texinfo[^0-9]*(4\.([4-9]|[1-9][0-9])|[5-9]|[1-9][0-9])' >/dev/null 2>&1; then
       :
     else
       MAKEINFO="$MISSING makeinfo"