diff -rup binutils-2.17.50.0.9/bfd/configure binutils-2.17.50.0.9-uclibc/bfd/configure
--- binutils-2.17.50.0.9/bfd/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/bfd/configure	2007-12-06 22:45:51.000000000 +0100
@@ -3579,6 +3579,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -rup binutils-2.17.50.0.9/binutils/configure binutils-2.17.50.0.9-uclibc/binutils/configure
--- binutils-2.17.50.0.9/binutils/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/binutils/configure	2007-12-06 22:46:37.000000000 +0100
@@ -3414,6 +3414,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -rup binutils-2.17.50.0.9/configure binutils-2.17.50.0.9-uclibc/configure
--- binutils-2.17.50.0.9/configure	2007-12-06 22:33:25.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/configure	2007-12-06 22:32:02.000000000 +0100
@@ -1438,6 +1438,18 @@ case "${target}" in
   i[3456789]86-*-coff | i[3456789]86-*-elf)
     noconfigdirs="$noconfigdirs ${libgcj}"
     ;;
+  i[3456789]86-*-linux-uclibc*)
+    # This section makes it possible to build newlib natively on linux.
+    # If we are using a cross compiler then don't configure newlib.
+    if test x${is_cross_compiler} != xno ; then
+      noconfigdirs="$noconfigdirs target-newlib"
+    fi
+    noconfigdirs="$noconfigdirs target-libgloss"
+    # If we are not using a cross compiler, do configure newlib.
+    # Note however, that newlib will only be configured in this situation
+    # if the --with-newlib option has been given, because otherwise
+    # 'target-newlib' will appear in skipdirs.
+    ;;
   i[3456789]86-*-linux*)
     # The GCC port for glibc1 has no MD_FALLBACK_FRAME_STATE_FOR, so let's
     # not build java stuff by default.
diff -rup binutils-2.17.50.0.9/configure.in binutils-2.17.50.0.9-uclibc/configure.in
--- binutils-2.17.50.0.9/configure.in	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/configure.in	2007-12-06 22:32:02.000000000 +0100
@@ -595,6 +595,18 @@ case "${target}" in
   i[[3456789]]86-*-coff | i[[3456789]]86-*-elf)
     noconfigdirs="$noconfigdirs ${libgcj}"
     ;;
+  i[[3456789]]86-*-linux-uclibc*)
+    # This section makes it possible to build newlib natively on linux.
+    # If we are using a cross compiler then don't configure newlib.
+    if test x${is_cross_compiler} != xno ; then
+      noconfigdirs="$noconfigdirs target-newlib"
+    fi
+    noconfigdirs="$noconfigdirs target-libgloss"
+    # If we are not using a cross compiler, do configure newlib.
+    # Note however, that newlib will only be configured in this situation
+    # if the --with-newlib option has been given, because otherwise
+    # 'target-newlib' will appear in skipdirs.
+    ;;
   i[[3456789]]86-*-linux*)
     # The GCC port for glibc1 has no MD_FALLBACK_FRAME_STATE_FOR, so let's
     # not build java stuff by default.
diff -rup binutils-2.17.50.0.9/gas/configure binutils-2.17.50.0.9-uclibc/gas/configure
--- binutils-2.17.50.0.9/gas/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/gas/configure	2007-12-06 22:49:50.000000000 +0100
@@ -3414,6 +3414,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -rup binutils-2.17.50.0.9/gprof/configure binutils-2.17.50.0.9-uclibc/gprof/configure
--- binutils-2.17.50.0.9/gprof/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/gprof/configure	2007-12-06 22:50:42.000000000 +0100
@@ -3413,6 +3413,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -rup binutils-2.17.50.0.9/ld/configure binutils-2.17.50.0.9-uclibc/ld/configure
--- binutils-2.17.50.0.9/ld/configure	2007-01-03 20:49:14.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/ld/configure	2007-12-06 22:51:12.000000000 +0100
@@ -3416,6 +3416,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -rup binutils-2.17.50.0.9/libtool.m4 binutils-2.17.50.0.9-uclibc/libtool.m4
--- binutils-2.17.50.0.9/libtool.m4	2006-02-14 00:10:15.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/libtool.m4	2007-12-06 22:32:03.000000000 +0100
@@ -743,6 +743,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     [lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$']
diff -rup binutils-2.17.50.0.9/ltconfig binutils-2.17.50.0.9-uclibc/ltconfig
--- binutils-2.17.50.0.9/ltconfig	2006-07-15 20:23:48.000000000 +0200
+++ binutils-2.17.50.0.9-uclibc/ltconfig	2007-12-06 22:32:03.000000000 +0100
@@ -603,6 +603,7 @@ host_os=`echo $host | sed 's/^\([^-]*\)-
 # Transform linux* to *-*-linux-gnu*, to support old configure scripts.
 case $host_os in
 linux-gnu*) ;;
+linux-uclibc*) ;;
 linux*) host=`echo $host | sed 's/^\(.*-.*-linux\)\(.*\)$/\1-gnu\2/'`
 esac
 
@@ -1270,6 +1271,24 @@ linux-gnu*)
   dynamic_linker='GNU/Linux ld.so'
   ;;
 
+linux-uclibc*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}.so$versuffix ${libname}${release}.so$major $libname.so'
+  soname_spec='${libname}${release}.so$major'
+  finish_cmds='PATH="\$PATH:/sbin" ldconfig -n $libdir'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=no
+  # This implies no fast_install, which is unacceptable.
+  # Some rework will be needed to allow for fast_install
+  # before this can be enabled.
+  # Note: copied from linux-gnu, and may not be appropriate.
+  hardcode_into_libs=yes
+  # Assume using the uClibc dynamic linker.
+  dynamic_linker="uClibc ld.so"
+  ;;
+
 netbsd*)
   need_lib_prefix=no
   need_version=no
diff -rup binutils-2.17.50.0.9/opcodes/configure binutils-2.17.50.0.9-uclibc/opcodes/configure
--- binutils-2.17.50.0.9/opcodes/configure	2006-11-27 23:21:04.000000000 +0100
+++ binutils-2.17.50.0.9-uclibc/opcodes/configure	2007-12-06 22:53:51.000000000 +0100
@@ -3582,6 +3582,11 @@ linux-gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
