diff -uNr binutils-2.17.50.0.9.org/bfd/Makefile.in binutils-2.17.50.0.9/bfd/Makefile.in
--- binutils-2.17.50.0.9.org/bfd/Makefile.in	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/bfd/Makefile.in	2009-04-01 16:21:29.000000000 +0200
@@ -256,7 +256,7 @@
 INCDIR = $(srcdir)/../include
 CSEARCH = -I. -I$(srcdir) -I$(INCDIR)
 MKDEP = gcc -MM
-SUBDIRS = doc po
+SUBDIRS = 
 bfddocdir = doc
 bfdlib_LTLIBRARIES = libbfd.la
 AM_CFLAGS = $(WARN_CFLAGS)
diff -uNr binutils-2.17.50.0.9.org/binutils/Makefile.in binutils-2.17.50.0.9/binutils/Makefile.in
--- binutils-2.17.50.0.9.org/binutils/Makefile.in	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/binutils/Makefile.in	2009-04-01 16:26:39.000000000 +0200
@@ -333,7 +333,7 @@
 target_os = @target_os@
 target_vendor = @target_vendor@
 AUTOMAKE_OPTIONS = cygnus dejagnu
-SUBDIRS = doc po
+SUBDIRS = 
 tooldir = $(exec_prefix)/$(target_alias)
 AM_CFLAGS = $(WARN_CFLAGS)
 
diff -uNr binutils-2.17.50.0.9.org/configure binutils-2.17.50.0.9/configure
--- binutils-2.17.50.0.9.org/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/configure	2009-04-01 16:46:35.000000000 +0200
@@ -921,7 +921,7 @@
 # know that we are building the simulator.
 # binutils, gas and ld appear in that order because it makes sense to run
 # "make check" in that particular order.
-host_tools="texinfo byacc flex bison binutils gas ld fixincludes gcc sid sim gdb make patch prms send-pr gprof etc expect dejagnu ash bash bzip2 m4 autoconf automake libtool diff rcs fileutils shellutils time textutils wdiff find uudecode hello tar gzip indent recode release sed utils guile perl gawk findutils gettext zip fastjar gnattools"
+host_tools="texinfo byacc flex bison binutils gas ld fixincludes gcc sid sim gdb make patch prms send-pr gprof expect dejagnu ash bash bzip2 m4 autoconf automake libtool diff rcs fileutils shellutils time textutils wdiff find uudecode hello tar gzip indent recode release sed utils guile perl gawk findutils gettext zip fastjar gnattools"
 
 # libgcj represents the runtime libraries only used by gcj.
 libgcj="target-libffi \
diff -uNr binutils-2.17.50.0.9.org/gas/Makefile.in binutils-2.17.50.0.9/gas/Makefile.in
--- binutils-2.17.50.0.9.org/gas/Makefile.in	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/gas/Makefile.in	2009-04-01 16:53:08.000000000 +0200
@@ -250,7 +250,7 @@
 target_vendor = @target_vendor@
 te_file = @te_file@
 AUTOMAKE_OPTIONS = 1.8 cygnus dejagnu
-SUBDIRS = doc po
+SUBDIRS = 
 # Automake should figure this out on its own.  It doesn't, because
 # of the "cygnus" option.  But distclean still wants it.
 DIST_SUBDIRS = $(SUBDIRS)
diff -uNr binutils-2.17.50.0.9.org/gprof/Makefile.in binutils-2.17.50.0.9/gprof/Makefile.in
--- binutils-2.17.50.0.9.org/gprof/Makefile.in	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.9/gprof/Makefile.in	2009-04-01 17:21:36.000000000 +0200
@@ -87,7 +87,7 @@
 LINK = $(LIBTOOL) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
 	$(AM_LDFLAGS) $(LDFLAGS) -o $@
 SOURCES = $(gprof_SOURCES)
-INFO_DEPS = gprof.info
+INFO_DEPS = 
 TEXINFO_TEX = $(top_srcdir)/../texinfo/texinfo.tex
 am__TEXINFO_TEX_DIR = $(top_srcdir)/../texinfo
 DVIS = gprof.dvi
diff -uNr binutils-2.17.50.0.9.org/ld/Makefile.in binutils-2.17.50.0.9/ld/Makefile.in
--- binutils-2.17.50.0.9.org/ld/Makefile.in	2007-01-03 20:49:14.000000000 +0100
+++ binutils-2.17.50.0.9/ld/Makefile.in	2009-04-01 18:48:09.000000000 +0200
@@ -109,7 +109,7 @@
 AM_MAKEINFOHTMLFLAGS = $(AM_MAKEINFOFLAGS)
 DVIPS = dvips
 RECURSIVE_TARGETS = all-recursive check-recursive dvi-recursive \
-	html-recursive info-recursive install-data-recursive \
+	html-recursive install-data-recursive \
 	install-exec-recursive install-info-recursive \
 	install-recursive installcheck-recursive installdirs-recursive \
 	pdf-recursive ps-recursive uninstall-info-recursive \
@@ -1181,8 +1181,6 @@
 
 install-exec-am: install-exec-local
 
-install-info: install-info-recursive
-
 install-info-am: $(INFO_DEPS)
 	@$(NORMAL_INSTALL)
 	test -z "$(infodir)" || $(mkdir_p) "$(DESTDIR)$(infodir)"
