? diff
? diff-old
Index: include/libbb.h
===================================================================
RCS file: /var/cvs/busybox/include/libbb.h,v
retrieving revision 1.91
diff -u -r1.91 libbb.h
--- include/libbb.h	13 Jan 2003 22:09:47 -0000	1.91
+++ include/libbb.h	19 Jan 2003 20:26:11 -0000
@@ -86,11 +86,11 @@
 extern void show_usage(void) __attribute__ ((noreturn));
 extern void error_msg(const char *s, ...) __attribute__ ((format (printf, 1, 2)));
 extern void error_msg_and_die(const char *s, ...) __attribute__ ((noreturn, format (printf, 1, 2)));
-extern void perror_msg(const char *s, ...);
-extern void perror_msg_and_die(const char *s, ...) __attribute__ ((noreturn));
+extern void perror_msg(const char *s, ...) __attribute__ ((format (printf, 1, 2)));
+extern void perror_msg_and_die(const char *s, ...) __attribute__ ((noreturn, format (printf, 1, 2)));
 extern void vherror_msg(const char *s, va_list p);
-extern void herror_msg(const char *s, ...);
-extern void herror_msg_and_die(const char *s, ...) __attribute__ ((noreturn));
+extern void herror_msg(const char *s, ...) __attribute__ ((format (printf, 1, 2)));
+extern void herror_msg_and_die(const char *s, ...) __attribute__ ((noreturn, format (printf, 1, 2)));
 
 /* These two are used internally -- you shouldn't need to use them */
 extern void verror_msg(const char *s, va_list p);
@@ -277,7 +277,6 @@
 extern const char * const group_file;
 extern const char * const securetty_file;
 extern const char * const motd_file;
-extern const char * const issue_file;
 extern const char * const _path_login;
 
 #ifdef CONFIG_FEATURE_DEVFS
@@ -364,5 +363,8 @@
 	struct llist_s *link;
 } llist_t;
 extern llist_t *llist_add_to(llist_t *old_head, char *new_item);
+
+void print_login_issue(const char *issue_file, const char *tty);
+void print_login_prompt(void);
 
 #endif /* __LIBCONFIG_H__ */
Index: libbb/Makefile.in
===================================================================
RCS file: /var/cvs/busybox/libbb/Makefile.in,v
retrieving revision 1.15
diff -u -r1.15 Makefile.in
--- libbb/Makefile.in	14 Dec 2002 01:58:58 -0000	1.15
+++ libbb/Makefile.in	19 Jan 2003 20:26:11 -0000
@@ -33,7 +33,7 @@
 	get_last_path_component.c get_line_from_file.c herror_msg.c \
 	herror_msg_and_die.c human_readable.c inet_common.c inode_hash.c \
 	interface.c isdirectory.c kernel_version.c last_char_is.c libc5.c \
-	llist_add_to.c loop.c make_directory.c mode_string.c \
+	llist_add_to.c login.c loop.c make_directory.c mode_string.c \
 	module_syscalls.c mtab.c mtab_file.c my_getgrgid.c my_getgrnam.c \
 	my_getpwnam.c my_getpwnamegid.c my_getpwuid.c obscure.c parse_mode.c \
 	parse_number.c perror_msg.c perror_msg_and_die.c print_file.c \
Index: libbb/login.c
===================================================================
RCS file: libbb/login.c
diff -N libbb/login.c
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ libbb/login.c	19 Jan 2003 20:26:11 -0000
@@ -0,0 +1,124 @@
+/*
+ * issue.c: issue printing code
+ *
+ * Copyright (C) 2003 Bastian Blank <waldi@tuxbox.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *
+ * $Id$
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+#include "busybox.h"
+
+#include <sys/utsname.h>
+#include <time.h>
+
+#define LOGIN " login: "
+
+static char fmtstr_d[] = { "%A, %d %B %Y" };
+static char fmtstr_t[] = { "%H:%M:%S" };
+
+void print_login_issue(const char *issue_file, const char *tty)
+{
+	FILE *fd;
+	int c;
+	char buf[256];
+	char *ret;
+	time_t t;
+	struct utsname uts;
+
+	time(&t);
+	uname(&uts);
+
+	puts("");	/* start a new line */
+
+	if ((fd = fopen(issue_file, "r"))) {
+		while ((c = fgetc(fd)) != EOF) {
+			if (c == '\\' || c == '%') {
+				c = fgetc(fd);
+
+				switch (c) {
+					case 's':
+						fputs(uts.sysname, stdout);
+						break;
+
+					case 'n':
+						fputs(uts.nodename, stdout);
+						break;
+
+					case 'r':
+						fputs(uts.release, stdout);
+						break;
+
+					case 'v':
+						fputs(uts.version, stdout);
+						break;
+
+					case 'm':
+						fputs(uts.machine, stdout);
+						break;
+
+					case 'D':
+					case 'o':
+						getdomainname(buf, sizeof(buf));
+						buf[sizeof(buf) - 1] = '\0';
+						fputs(buf, stdout);
+						break;
+
+					case 'd':
+						strftime(buf, sizeof(buf), fmtstr_d, localtime(&t));
+						fputs(buf, stdout);
+						break;
+
+					case 't':
+						strftime(buf, sizeof(buf), fmtstr_t, localtime(&t));
+						fputs(buf, stdout);
+						break;
+
+					case 'h':
+						gethostname(buf, sizeof(buf));
+						fputs(buf, stdout);
+						break;
+
+					case 'l':
+						printf("%s", tty);
+
+					default:
+						putchar(c);
+				}
+			} else
+				putchar(c);
+		}
+
+		puts("");	/* start a new line */
+		fflush(stdout);
+
+		fclose(fd);
+	}
+}
+
+void print_login_prompt(void)
+{
+	char buf[MAXHOSTNAMELEN];
+
+	gethostname(buf, MAXHOSTNAMELEN);
+	fputs(buf, stdout);
+
+	fputs(LOGIN, stdout);
+	fflush(stdout);
+}
+
Index: libbb/run_parts.c
===================================================================
RCS file: /var/cvs/busybox/libbb/run_parts.c,v
retrieving revision 1.3
diff -u -r1.3 run_parts.c
--- libbb/run_parts.c	21 Nov 2002 22:17:11 -0000	1.3
+++ libbb/run_parts.c	19 Jan 2003 20:26:11 -0000
@@ -64,6 +64,8 @@
 	entries = scandir(arg0, &namelist, valid_name, alphasort);
 
 	if (entries == -1) {
+		if (test_mode & 2)
+			return 2;
 		perror_msg_and_die("failed to open directory %s", arg0);
 	}
 
@@ -75,7 +77,7 @@
 			perror_msg_and_die("failed to stat component %s", filename);
 		}
 		if (S_ISREG(st.st_mode) && !access(filename, X_OK)) {
-			if (test_mode)
+			if (test_mode & 1)
 				printf("run-parts would run %s\n", filename);
 			else {
 				/* exec_errno is common vfork variable */
Index: loginutils/Config.in
===================================================================
RCS file: /var/cvs/busybox/loginutils/Config.in,v
retrieving revision 1.1
diff -u -r1.1 Config.in
--- loginutils/Config.in	5 Dec 2002 08:40:47 -0000	1.1
+++ loginutils/Config.in	19 Jan 2003 20:26:12 -0000
@@ -54,7 +54,6 @@
 	help
 	  Please submit a patch to add help text for this item.
 
-
 config CONFIG_PASSWD
 	bool "passwd"
 	default n
@@ -67,29 +66,32 @@
 	help
 	  Please submit a patch to add help text for this item.
 
-config CONFIG_FEATURE_SHADOWPASSWDS
-	bool "Support for shadow passwords"
-	default y
-	depends on CONFIG_ADDUSER || CONFIG_DELUSER || CONFIG_LOGIN || CONFIG_SU
+config CONFIG_SULOGIN
+	bool "sulogin"
+	default n
 	help
 	  Please submit a patch to add help text for this item.
 
-config CONFIG_USE_BB_SHADOW
-	bool "  Use busybox shadow password functions"
+config CONFIG_VLOCK
+	bool "vlock"
 	default n
-	depends on CONFIG_USE_BB_PWD_GRP && CONFIG_FEATURE_SHADOWPASSWDS
 	help
 	  Please submit a patch to add help text for this item.
 
-config CONFIG_SULOGIN
-	bool "sulogin"
+comment "Common options for adduser, deluser, login, su"
+	depends on CONFIG_ADDUSER || CONFIG_DELUSER || CONFIG_LOGIN || CONFIG_SU
+
+config CONFIG_FEATURE_SHADOWPASSWDS
+	bool "Support for shadow passwords"
 	default n
+	depends on CONFIG_ADDUSER || CONFIG_DELUSER || CONFIG_LOGIN || CONFIG_SU
 	help
 	  Please submit a patch to add help text for this item.
 
-config CONFIG_VLOCK
-	bool "vlock"
+config CONFIG_USE_BB_SHADOW
+	bool "  Use busybox shadow password functions"
 	default n
+	depends on CONFIG_USE_BB_PWD_GRP && CONFIG_FEATURE_SHADOWPASSWDS
 	help
 	  Please submit a patch to add help text for this item.
 
Index: loginutils/getty.c
===================================================================
RCS file: /var/cvs/busybox/loginutils/getty.c,v
retrieving revision 1.5
diff -u -r1.5 getty.c
--- loginutils/getty.c	10 Oct 2002 04:20:17 -0000	1.5
+++ loginutils/getty.c	19 Jan 2003 20:26:12 -0000
@@ -73,8 +73,6 @@
 #include <time.h>
 #endif
 
-#define LOGIN " login: "		/* login prompt */
-
 /* Some shorthands for control characters. */
 
 #define CTL(x)		(x ^ 0100)	/* Assumes ASCII dialect */
@@ -752,142 +750,10 @@
 /* do_prompt - show login prompt, optionally preceded by /etc/issue contents */
 static void do_prompt(struct options *op, struct termio *tp)
 {
-#ifdef	ISSUE
-	FILE *fd;
-	int oflag;
-	int c;
-	struct utsname uts;
-
-	(void) uname(&uts);
-#endif
-
-	(void) write(1, "\r\n", 2);	/* start a new line */
 #ifdef	ISSUE					/* optional: show /etc/issue */
-	if ((op->flags & F_ISSUE) && (fd = fopen(op->issue, "r"))) {
-		oflag = tp->c_oflag;	/* save current setting */
-		tp->c_oflag |= (ONLCR | OPOST);	/* map NL in output to CR-NL */
-		(void) ioctl(0, TCSETAW, tp);
-
-
-		while ((c = getc(fd)) != EOF) {
-			if (c == '\\') {
-				c = getc(fd);
-
-				switch (c) {
-				case 's':
-					(void) printf("%s", uts.sysname);
-					break;
-
-				case 'n':
-					(void) printf("%s", uts.nodename);
-					break;
-
-				case 'r':
-					(void) printf("%s", uts.release);
-					break;
-
-				case 'v':
-					(void) printf("%s", uts.version);
-					break;
-
-				case 'm':
-					(void) printf("%s", uts.machine);
-					break;
-
-				case 'o':
-				{
-					char domainname[256];
-
-					getdomainname(domainname, sizeof(domainname));
-					domainname[sizeof(domainname) - 1] = '\0';
-					printf("%s", domainname);
-				}
-					break;
-
-				case 'd':
-				case 't':
-				{
-					char *weekday[] = { "Sun", "Mon", "Tue", "Wed", "Thu",
-						"Fri", "Sat"
-					};
-					char *month[] = { "Jan", "Feb", "Mar", "Apr", "May",
-						"Jun", "Jul", "Aug", "Sep", "Oct",
-						"Nov", "Dec"
-					};
-					time_t now;
-					struct tm *tm;
-
-					(void) time(&now);
-					tm = localtime(&now);
-
-					if (c == 'd')
-						(void) printf("%s %s %d  %d",
-									  weekday[tm->tm_wday],
-									  month[tm->tm_mon], tm->tm_mday,
-									  tm->tm_year <
-									  70 ? tm->tm_year +
-									  2000 : tm->tm_year + 1900);
-					else
-						(void) printf("%02d:%02d:%02d", tm->tm_hour,
-									  tm->tm_min, tm->tm_sec);
-
-					break;
-				}
-
-				case 'l':
-					(void) printf("%s", op->tty);
-					break;
-
-				case 'b':
-				{
-					int i;
-
-					for (i = 0; speedtab[i].speed; i++) {
-						if (speedtab[i].code == (tp->c_cflag & CBAUD)) {
-							printf("%ld", speedtab[i].speed);
-							break;
-						}
-					}
-					break;
-				}
-				case 'u':
-				case 'U':
-				{
-					int users = 0;
-					struct utmp *ut;
-
-					setutent();
-					while ((ut = getutent()))
-						if (ut->ut_type == USER_PROCESS)
-							users++;
-					endutent();
-					printf("%d ", users);
-					if (c == 'U')
-						printf((users == 1) ? "user" : "users");
-					break;
-				}
-				default:
-					(void) putchar(c);
-				}
-			} else
-				(void) putchar(c);
-		}
-		fflush(stdout);
-
-		tp->c_oflag = oflag;	/* restore settings */
-		(void) ioctl(0, TCSETAW, tp);	/* wait till output is gone */
-		(void) fclose(fd);
-	}
-#endif
-#ifdef __linux__
-	{
-		char hn[MAXHOSTNAMELEN + 1];
-
-		(void) gethostname(hn, MAXHOSTNAMELEN);
-		write(1, hn, strlen(hn));
-	}
+	print_login_issue(op->issue, op->tty);
 #endif
-	(void) write(1, LOGIN, sizeof(LOGIN) - 1);	/* always show login prompt */
+	print_login_prompt();
 }
 
 /* next_speed - select next baud rate */
Index: loginutils/login.c
===================================================================
RCS file: /var/cvs/busybox/loginutils/login.c,v
retrieving revision 1.8
diff -u -r1.8 login.c
--- loginutils/login.c	3 Dec 2002 19:54:12 -0000	1.8
+++ loginutils/login.c	19 Jan 2003 20:26:13 -0000
@@ -253,20 +253,18 @@
 	int i;
 
 	for(i=0; i<EMPTY_USERNAME_COUNT; i++) {
-	gethostname ( buf, sizeof( buf ));
-	printf ( "\n%s login: ", buf );
-	fflush ( stdout );
+		print_login_prompt();
+
+		if ( !fgets ( buf, sizeof( buf ) - 1, stdin ))
+			return 0;
 
-	if ( !fgets ( buf, sizeof( buf ) - 1, stdin ))
-		return 0;
-		
 		if ( !strchr ( buf, '\n' ))
-		return 0;
-	
-	for ( sp = buf; isspace ( *sp ); sp++ ) { }
-	for ( ep = sp; isgraph ( *ep ); ep++ ) { }
+			return 0;
+
+		for ( sp = buf; isspace ( *sp ); sp++ ) { }
+		for ( ep = sp; isgraph ( *ep ); ep++ ) { }
 
-	*ep = 0;		
+		*ep = 0;		
 		safe_strncpy(buf_name, sp, USERNAME_SIZE);
 		if(buf_name[0])
 			return 1;
Index: networking/Config.in
===================================================================
RCS file: /var/cvs/busybox/networking/Config.in,v
retrieving revision 1.8
diff -u -r1.8 Config.in
--- networking/Config.in	13 Jan 2003 21:40:37 -0000	1.8
+++ networking/Config.in	19 Jan 2003 20:26:13 -0000
@@ -315,6 +315,13 @@
 	help
 	  Please submit a patch to add help text for this item.
 
+config CONFIG_FEATURE_TELNETD_INETD
+	bool "  Use inetd"
+	default n
+	depends on CONFIG_TELNETD
+	help
+	  Please submit a patch to add help text for this item.
+
 config CONFIG_TFTP
 	bool "tftp"
 	default n
Index: networking/ifupdown.c
===================================================================
RCS file: /var/cvs/busybox/networking/ifupdown.c,v
retrieving revision 1.17
diff -u -r1.17 ifupdown.c
--- networking/ifupdown.c	14 Jan 2003 23:26:57 -0000	1.17
+++ networking/ifupdown.c	19 Jan 2003 20:26:13 -0000
@@ -1010,7 +1010,7 @@
 
 	buf = xmalloc(xstrlen(opt) + 19);
 	sprintf(buf, "/etc/network/if-%s.d", opt);
-	run_parts(&buf, 0);
+	run_parts(&buf, 2);
 	free(buf);
 	return (1);
 }
Index: networking/telnetd.c
===================================================================
RCS file: /var/cvs/busybox/networking/telnetd.c,v
retrieving revision 1.2
diff -u -r1.2 telnetd.c
--- networking/telnetd.c	10 Nov 2002 22:26:19 -0000	1.2
+++ networking/telnetd.c	19 Jan 2003 20:26:14 -0000
@@ -48,7 +48,13 @@
 
 #define BUFSIZE 4000
 
-static const char *loginpath = "/bin/sh";
+static const char *loginpath = 
+#ifdef CONFIG_LOGIN
+"/bin/login";
+#else
+"/bin/sh";
+#endif
+static const char *issuefile = "/etc/issue.net";
 
 /* shell name and arguments */
 
@@ -57,8 +63,12 @@
 /* structure that describes a session */
 
 struct tsession {
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+	int sockfd_read, sockfd_write, ptyfd;
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 	struct tsession *next;
 	int sockfd, ptyfd;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 	int shell_pid;
 	/* two circular buffers */
 	char *buf1, *buf2;
@@ -204,7 +214,11 @@
 
 
 static struct tsession *
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+make_new_session(void)
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 make_new_session(int sockfd)
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 {
 	struct termios termbuf;
 	int pty, pid;
@@ -214,7 +228,12 @@
 	ts->buf1 = (char *)(&ts[1]);
 	ts->buf2 = ts->buf1 + BUFSIZE;
 
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+	ts->sockfd_read = 0;
+	ts->sockfd_write = 1;
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 	ts->sockfd = sockfd;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 	ts->rdidx1 = ts->wridx1 = ts->size1 = 0;
 	ts->rdidx2 = ts->wridx2 = ts->size2 = 0;
@@ -278,6 +297,8 @@
 		/*termbuf.c_lflag &= ~ICANON;*/
 		tcsetattr(0, TCSANOW, &termbuf);
 
+		print_login_issue(issuefile, NULL);
+
 		/* exec shell, with correct argv and env */
 		execv(loginpath, (char *const *)argv_init);
 
@@ -291,6 +312,7 @@
 	return ts;
 }
 
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 static void
 free_session(struct tsession *ts)
 {
@@ -319,30 +341,44 @@
 
 	free(ts);
 }
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 int
 telnetd_main(int argc, char **argv)
 {
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 	struct sockaddr_in sa;
 	int master_fd;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 	fd_set rdfdset, wrfdset;
 	int selret;
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 	int on = 1;
 	int portnbr = 23;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 	int c;
-
-	/* check if user supplied a port number */
+	static const char options[] =
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+		"f:l:";
+#else /* CONFIG_EATURE_TELNETD_INETD */
+		"f:l:p:";
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 	for (;;) {
-		c = getopt( argc, argv, "p:l:");
+		c = getopt( argc, argv, options);
 		if (c == EOF) break;
 		switch (c) {
-			case 'p':
-				portnbr = atoi(optarg);
+			case 'f':
+				issuefile = strdup (optarg);
 				break;
 			case 'l':
 				loginpath = strdup (optarg);
 				break;
+#ifndef CONFIG_FEATURE_TELNETD_INETD
+			case 'p':
+				portnbr = atoi(optarg);
+				break;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 			default:
 				show_usage();
 		}
@@ -353,6 +389,12 @@
 	}
 
 	argv_init[0] = loginpath;
+
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+	sessions = make_new_session();
+
+	maxfd = 1;
+#else /* CONFIG_EATURE_TELNETD_INETD */
 	sessions = 0;
 
 	/* Grab a TCP socket.  */
@@ -382,6 +424,7 @@
 
 
 	maxfd = master_fd;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 	do {
 		struct tsession *ts;
@@ -393,10 +436,14 @@
 		 * ptys if there is room in their respective session buffers.
 		 */
 
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 		FD_SET(master_fd, &rdfdset);
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 		ts = sessions;
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 		while (ts) {
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 			/* buf1 is used from socket to pty
 			 * buf2 is used from pty to socket
 			 */
@@ -404,22 +451,33 @@
 				FD_SET(ts->ptyfd, &wrfdset);  /* can write to pty */
 			}
 			if (ts->size1 < BUFSIZE) {
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+				FD_SET(ts->sockfd_read, &rdfdset); /* can read from socket */
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 				FD_SET(ts->sockfd, &rdfdset); /* can read from socket */
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 			}
 			if (ts->size2 > 0) {
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+				FD_SET(ts->sockfd_write, &wrfdset); /* can write to socket */
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 				FD_SET(ts->sockfd, &wrfdset); /* can write to socket */
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 			}
 			if (ts->size2 < BUFSIZE) {
 				FD_SET(ts->ptyfd, &rdfdset);  /* can read from pty */
 			}
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 			ts = ts->next;
 		}
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 		selret = select(maxfd + 1, &rdfdset, &wrfdset, 0, 0);
 
 		if (!selret)
 			break;
 
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 		/* First check for and accept new sessions.  */
 		if (FD_ISSET(master_fd, &rdfdset)) {
 			int fd, salen;
@@ -447,9 +505,12 @@
 
 		ts = sessions;
 		while (ts) { /* For all sessions...  */
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 			int maxlen, w, r;
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 			struct tsession *next = ts->next; /* in case we free ts. */
-
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
+			
 			if (ts->size1 && FD_ISSET(ts->ptyfd, &wrfdset)) {
 				int num_totty;
 				char *ptr;
@@ -459,9 +520,13 @@
 
 				w = write(ts->ptyfd, ptr, num_totty);
 				if (w < 0) {
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+					exit(0);
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 					free_session(ts);
 					ts = next;
 					continue;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				}
 				ts->wridx1 += w;
 				ts->size1 -= w;
@@ -469,31 +534,51 @@
 					ts->wridx1 = 0;
 			}
 
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+			if (ts->size2 && FD_ISSET(ts->sockfd_write, &wrfdset)) {
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 			if (ts->size2 && FD_ISSET(ts->sockfd, &wrfdset)) {
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				/* Write to socket from buffer 2.  */
 				maxlen = MIN(BUFSIZE - ts->wridx2, ts->size2);
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+				w = write(ts->sockfd_write, ts->buf2 + ts->wridx2, maxlen);
+				if (w < 0)
+					exit(0);
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 				w = write(ts->sockfd, ts->buf2 + ts->wridx2, maxlen);
 				if (w < 0) {
 					free_session(ts);
 					ts = next;
 					continue;
 				}
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				ts->wridx2 += w;
 				ts->size2 -= w;
 				if (ts->wridx2 == BUFSIZE)
 					ts->wridx2 = 0;
 			}
 
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+			if (ts->size1 < BUFSIZE && FD_ISSET(ts->sockfd_read, &rdfdset)) {
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 			if (ts->size1 < BUFSIZE && FD_ISSET(ts->sockfd, &rdfdset)) {
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				/* Read from socket to buffer 1. */
 				maxlen = MIN(BUFSIZE - ts->rdidx1,
 						BUFSIZE - ts->size1);
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+				r = read(ts->sockfd_read, ts->buf1 + ts->rdidx1, maxlen);
+				if (!r || (r < 0 && errno != EINTR))
+					exit(0);
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 				r = read(ts->sockfd, ts->buf1 + ts->rdidx1, maxlen);
 				if (!r || (r < 0 && errno != EINTR)) {
 					free_session(ts);
 					ts = next;
 					continue;
 				}
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				if(!*(ts->buf1 + ts->rdidx1 + r - 1)) {
 					r--;
 					if(!r)
@@ -511,9 +596,13 @@
 						BUFSIZE - ts->size2);
 				r = read(ts->ptyfd, ts->buf2 + ts->rdidx2, maxlen);
 				if (!r || (r < 0 && errno != EINTR)) {
+#ifdef CONFIG_FEATURE_TELNETD_INETD
+					exit(0);
+#else /* CONFIG_FEATURE_TELNETD_INETD */
 					free_session(ts);
 					ts = next;
 					continue;
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 				}
 				ts->rdidx2 += r;
 				ts->size2 += r;
@@ -529,8 +618,10 @@
 				ts->rdidx2 = 0;
 				ts->wridx2 = 0;
 			}
+#ifndef CONFIG_FEATURE_TELNETD_INETD
 			ts = next;
 		}
+#endif /* CONFIG_FEATURE_TELNETD_INETD */
 
 	} while (1);
 
Index: networking/libiproute/ip_common.h
===================================================================
RCS file: /var/cvs/busybox/networking/libiproute/ip_common.h,v
retrieving revision 1.4
diff -u -r1.4 ip_common.h
--- networking/libiproute/ip_common.h	2 Dec 2002 01:40:05 -0000	1.4
+++ networking/libiproute/ip_common.h	19 Jan 2003 20:26:14 -0000
@@ -3,7 +3,7 @@
 
 extern void ip_parse_common_args(int *argcp, char ***argvp);
 extern int print_neigh(struct sockaddr_nl *who, struct nlmsghdr *n, void *arg);
-extern int ipaddr_list(int argc, char **argv);
+extern int ipaddr_list_or_flush(int argc, char **argv, int flush);
 extern int iproute_monitor(int argc, char **argv);
 extern void iplink_usage(void) __attribute__((noreturn));
 extern void ipneigh_reset_filter(void);
Index: sysdeps/linux/Config.in
===================================================================
RCS file: /var/cvs/busybox/sysdeps/linux/Config.in,v
retrieving revision 1.4
retrieving revision 1.5
diff -u -r1.4 -r1.5
--- sysdeps/linux/Config.in	13 Dec 2002 22:53:27 -0000	1.4
+++ sysdeps/linux/Config.in	19 Jan 2003 12:55:13 -0000	1.5
@@ -157,6 +157,17 @@
 
 endmenu
 
+menu 'Installation Options'
+
+config CONFIG_INSTALL_NO_USR
+	bool "Don't use /usr"
+	default n
+	help
+	  Disable use of /usr. Don't activate this option if you don't know,
+	  that you really want this behaviour.
+
+endmenu
+
 source archival/Config.in
 source console-tools/Config.in
 source debianutils/Config.in
Index: include/applets.h
===================================================================
RCS file: /var/cvs/busybox/include/applets.h,v
retrieving revision 1.88
retrieving revision 1.89
diff -u -r1.88 -r1.89
--- include/applets.h	14 Jan 2003 21:41:12 -0000	1.88
+++ include/applets.h	19 Jan 2003 12:55:11 -0000	1.89
@@ -41,6 +41,11 @@
   #define APPLET_ODDNAME(a,b,c,d,e) {a,b,c,d},
 #endif
 
+#ifdef CONFIG_INSTALL_NO_USR
+#define _BB_DIR_USR_BIN _BB_DIR_BIN
+#define _BB_DIR_USR_SBIN _BB_DIR_SBIN
+#endif
+
 
 
 #ifdef CONFIG_TEST
