diff -ur busybox-1.1.0-original/init/init.c busybox-1.1.0/init/init.c
--- busybox-1.1.0-original/init/init.c	2006-01-11 06:44:17.000000000 +0100
+++ busybox-1.1.0/init/init.c	2006-03-04 23:23:07.000000000 +0100
@@ -344,7 +344,7 @@
 	while ((fd = open(console, O_RDONLY | O_NONBLOCK)) < 0 && tried < 2) {
 		/* Can't open selected console -- try
 			logical system console and VT_MASTER */
-		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : CURRENT_VC),
+		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : "/dev/tty0"),
 							sizeof(console));
 		tried++;
 	}
@@ -608,15 +608,12 @@
 
 	pid = run(a);
 	while (1) {
-		wpid = waitpid(pid,&status,0);
+		wpid = wait(&status);
+		if (wpid > 0 && wpid != pid) {
+			continue;
+		}	
 		if (wpid == pid)
 			break;
-		if (wpid == -1 && errno == ECHILD) {
-			/* we missed its termination */
-			break;
-		}
-		/* FIXME other errors should maybe trigger an error, but allow
-		 * the program to continue */
 	}
 	return wpid;
 }
diff -ur busybox-1.1.0-original/libbb/messages.c busybox-1.1.0/libbb/messages.c
--- busybox-1.1.0-original/libbb/messages.c	2006-01-11 06:44:06.000000000 +0100
+++ busybox-1.1.0/libbb/messages.c	2006-03-04 23:23:07.000000000 +0100
@@ -59,27 +59,47 @@
 #endif
 
 #ifdef L_passwd_file
-#define PASSWD_FILE        "/etc/passwd"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define PASSWD_FILE        "/var/etc/passwd"
+#else
+ #define PASSWD_FILE        "/etc/passwd"
+#endif
 const char * const bb_path_passwd_file = PASSWD_FILE;
 #endif
 
 #ifdef L_shadow_file
-#define SHADOW_FILE        "/etc/shadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define SHADOW_FILE        "/var/etc/shadow"
+#else
+ #define SHADOW_FILE        "/etc/shadow"
+#endif
 const char * const bb_path_shadow_file = SHADOW_FILE;
 #endif
 
 #ifdef L_group_file
-#define GROUP_FILE         "/etc/group"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GROUP_FILE         "/var/etc/group"
+#else
+ #define GROUP_FILE         "/etc/group"
+#endif
 const char * const bb_path_group_file = GROUP_FILE;
 #endif
 
 #ifdef L_gshadow_file
-#define GSHADOW_FILE       "/etc/gshadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GSHADOW_FILE       "/var/etc/gshadow"
+#else
+ #define GSHADOW_FILE       "/etc/gshadow"
+#endif
 const char * const bb_path_gshadow_file = GSHADOW_FILE;
 #endif
 
 #ifdef L_nologin_file
-#define NOLOGIN_FILE       "/etc/nologin"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define NOLOGIN_FILE       "/var/etc/nologin"
+#else
+ #define NOLOGIN_FILE       "/etc/nologin"
+#endif
 const char * const bb_path_nologin_file = NOLOGIN_FILE;
 #endif
 
diff -ur busybox-1.1.0-original/miscutils/Config.in busybox-1.1.0/miscutils/Config.in
--- busybox-1.1.0-original/miscutils/Config.in	2006-01-11 06:44:13.000000000 +0100
+++ busybox-1.1.0/miscutils/Config.in	2006-03-05 17:00:24.000000000 +0100
@@ -291,7 +291,7 @@
 	  When the command finishes, time writes a message to standard output
 	  giving timing statistics about this program run.
 
-config CONFIG_WATCHDOG
+config CONFIG_BB_WATCHDOG
 	bool "watchdog"
 	default n
 	help
Only in busybox-1.1.0/miscutils: Config.in~
diff -ur busybox-1.1.0-original/miscutils/Makefile.in busybox-1.1.0/miscutils/Makefile.in
--- busybox-1.1.0-original/miscutils/Makefile.in	2006-01-11 06:44:13.000000000 +0100
+++ busybox-1.1.0/miscutils/Makefile.in	2006-03-05 17:00:16.000000000 +0100
@@ -29,7 +29,7 @@
 MISCUTILS-$(CONFIG_SETSID)      += setsid.o
 MISCUTILS-$(CONFIG_STRINGS)     += strings.o
 MISCUTILS-$(CONFIG_TIME)        += time.o
-MISCUTILS-$(CONFIG_WATCHDOG)    += watchdog.o
+MISCUTILS-$(CONFIG_BB_WATCHDOG) += watchdog.o
 
 libraries-y+=$(MISCUTILS_DIR)$(MISCUTILS_AR)
 
diff -ur busybox-1.1.0-original/networking/ifupdown.c busybox-1.1.0/networking/ifupdown.c
--- busybox-1.1.0-original/networking/ifupdown.c	2006-03-22 22:16:19.000000000 +0100
+++ busybox-1.1.0/networking/ifupdown.c	2006-04-14 21:47:58.000000000 +0200
@@ -971,7 +971,7 @@
 
 static int execute_all(struct interface_defn_t *ifd, const char *opt)
 {
-	int i;
+	int i, fd;
 	char *buf;
 	for (i = 0; i < ifd->n_options; i++) {
 		if (strcmp(ifd->option[i].name, opt) == 0) {
@@ -981,9 +981,13 @@
 		}
 	}
 
-	buf = bb_xasprintf("run-parts /etc/network/if-%s.d", opt);
-	if (doit(buf) != 1) {
-		return 0;
+	buf = bb_xasprintf("/etc/network/if-%s.d", opt);
+	if ((fd = open(buf, O_RDONLY)) >= 0) {
+		close(fd);
+		buf = bb_xasprintf("run-parts /etc/network/if-%s.d", opt);
+		if (doit(buf) != 1) {
+			return 0;
+		}
 	}
 	return 1;
 }
diff -ur busybox-1.1.0-original/sysdeps/linux/Config.in busybox-1.1.0/sysdeps/linux/Config.in
--- busybox-1.1.0-original/Config.in	2006-01-11 06:44:17.000000000 +0100
+++ busybox-1.1.0/Config.in	2006-03-04 23:23:07.000000000 +0100
@@ -218,6 +218,14 @@
 	help
 	  Disables support for numeric arguments in fold.
 
+config CONFIG_AUTH_IN_VAR_ETC
+       bool "Search auth files in /var/etc"
+       default n
+       help
+         passwd, shadow, group, gshadow, and nologin will be searched in
+         /var/etc instead of /etc. This is usefull, if /etc is not
+         writeable (flash). 
+
 endmenu
 
 menu 'Installation Options'
diff -ur busybox-1.1.0-original/include/applets.h busybox-1.1.0/include/applets.h
--- busybox-1.1.0-original/include/applets.h	2006-03-22 22:16:24.000000000 +0100
+++ busybox-1.1.0/include/applets.h	2006-04-14 22:00:14.000000000 +0200
@@ -288,7 +288,7 @@
 USE_VI(APPLET(vi, vi_main, _BB_DIR_BIN, _BB_SUID_NEVER))
 USE_VLOCK(APPLET(vlock, vlock_main, _BB_DIR_USR_BIN, _BB_SUID_ALWAYS))
 USE_WATCH(APPLET(watch, watch_main, _BB_DIR_BIN, _BB_SUID_NEVER))
-USE_WATCHDOG(APPLET(watchdog, watchdog_main, _BB_DIR_SBIN, _BB_SUID_NEVER))
+USE_BB_WATCHDOG(APPLET(watchdog, watchdog_main, _BB_DIR_SBIN, _BB_SUID_NEVER))
 USE_WC(APPLET(wc, wc_main, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
 USE_WGET(APPLET(wget, wget_main, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
 USE_WHICH(APPLET(which, which_main, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
