diff -Naur busybox-1.00-pre7/init/init.c busybox-1.00-pre7-dbox2/init/init.c
--- busybox-1.00-pre7/init/init.c	2003-09-26 12:45:55.000000000 +0200
+++ busybox-1.00-pre7-dbox2/init/init.c	2004-02-22 16:13:00.000000000 +0100
@@ -376,7 +376,7 @@
 	while ((fd = open(console, O_RDONLY | O_NONBLOCK)) < 0 && tried < 2) {
 		/* Can't open selected console -- try
 			logical system console and VT_MASTER */
-		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : CURRENT_VC),
+		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : "/dev/tty0"),
 							sizeof(console));
 		tried++;
 	}
@@ -629,15 +629,12 @@
 
 	pid = run(a);
 	while (1) {
-		wpid = waitpid(pid,&status,0);
+		wpid = wait(&status);
+		if (wpid > 0 && wpid != pid) {
+			continue;
+		}
 		if (wpid == pid)
 			break;
-		if (wpid == -1 && errno == ECHILD) {
-			/* we missed its termination */
-			break;
-		}
-		/* FIXME other errors should maybe trigger an error, but allow
-		 * the program to continue */
 	}
 	return wpid;
 }
diff -Naur busybox-1.00-pre7/libbb/messages.c busybox-1.00-pre7-dbox2/libbb/messages.c
--- busybox-1.00-pre7/libbb/messages.c	2003-09-02 04:36:11.000000000 +0200
+++ busybox-1.00-pre7-dbox2/libbb/messages.c	2004-02-22 16:11:11.000000000 +0100
@@ -56,27 +56,47 @@
 #endif
 
 #ifdef L_passwd_file
-#define PASSWD_FILE        "/etc/passwd"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define PASSWD_FILE        "/var/etc/passwd"
+#else
+ #define PASSWD_FILE        "/etc/passwd"
+#endif
 const char * const bb_path_passwd_file = PASSWD_FILE;
 #endif
 
 #ifdef L_shadow_file
-#define SHADOW_FILE        "/etc/shadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define SHADOW_FILE        "/var/etc/shadow"
+#else
+ #define SHADOW_FILE        "/etc/shadow"
+#endif
 const char * const bb_path_shadow_file = SHADOW_FILE;
 #endif
 
 #ifdef L_group_file
-#define GROUP_FILE         "/etc/group"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GROUP_FILE         "/var/etc/group"
+#else
+ #define GROUP_FILE         "/etc/group"
+#endif
 const char * const bb_path_group_file = GROUP_FILE;
 #endif
 
 #ifdef L_gshadow_file
-#define GSHADOW_FILE       "/etc/gshadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GSHADOW_FILE       "/var/etc/gshadow"
+#else
+ #define GSHADOW_FILE       "/etc/gshadow"
+#endif
 const char * const bb_path_gshadow_file = GSHADOW_FILE;
 #endif
 
 #ifdef L_nologin_file
-#define NOLOGIN_FILE       "/etc/nologin"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define NOLOGIN_FILE       "/var/etc/nologin"
+#else
+ #define NOLOGIN_FILE       "/etc/nologin"
+#endif
 const char * const bb_path_nologin_file = NOLOGIN_FILE;
 #endif
 
diff -Naur busybox-1.00-pre7/networking/ifupdown.c busybox-1.00-pre7-dbox2/networking/ifupdown.c
--- busybox-1.00-pre7/networking/ifupdown.c	2003-12-19 11:46:00.000000000 +0100
+++ busybox-1.00-pre7-dbox2/networking/ifupdown.c	2004-02-22 16:11:11.000000000 +0100
@@ -1022,7 +1022,7 @@
 
 static int execute_all(struct interface_defn_t *ifd, execfn *exec, const char *opt)
 {
-	int i;
+	int i, fd;
 	char *buf;
 	for (i = 0; i < ifd->n_options; i++) {
 		if (strcmp(ifd->option[i].name, opt) == 0) {
@@ -1031,9 +1031,13 @@
 			}
 		}
 	}
-	
-	bb_xasprintf(&buf, "run-parts /etc/network/if-%s.d", opt);
-	(*exec)(buf);
+
+	bb_xasprintf(&buf, "/etc/network/if-%s.d", opt);
+	if ((fd = open(buf, O_RDONLY)) >= 0) {
+		close(fd);
+		bb_xasprintf(&buf, "run-parts /etc/network/if-%s.d", opt);
+		(*exec)(buf);
+	}
 
 	return (1);
 }
diff -Naur busybox-1.00-pre7/sysdeps/linux/Config.in busybox-1.00-pre7-dbox2/sysdeps/linux/Config.in
--- busybox-1.00-pre7/sysdeps/linux/Config.in	2003-10-22 11:58:50.000000000 +0200
+++ busybox-1.00-pre7-dbox2/sysdeps/linux/Config.in	2004-02-22 16:11:11.000000000 +0100
@@ -195,6 +195,14 @@
 	  if you want to add some simple compiler switches (like -march=i686),
 	  or check for warnings using -Werror, just those options here.
 
+config CONFIG_AUTH_IN_VAR_ETC
+	bool "Search auth files in /var/etc"
+	default n
+	help
+	  passwd, shadow, group, gshadow and nologin will be searched in
+	  /var/etc instead of /etc. This is usefull, if /etc is not
+	  writeable (flash).
+
 endmenu
 
 menu 'Installation Options'
