diff -urN busybox-cvs-0.60.99.cvs20030114.orig/include/applets.h busybox-cvs-0.60.99.cvs20030114/include/applets.h
--- busybox-cvs-0.60.99.cvs20030114.orig/include/applets.h	2003-01-13 18:26:23.000000000 +0100
+++ busybox-cvs-0.60.99.cvs20030114/include/applets.h	2003-01-16 20:47:12.000000000 +0100
@@ -41,7 +41,8 @@
   #define APPLET_ODDNAME(a,b,c,d,e) {a,b,c,d},
 #endif
 
-
+#define _BB_DIR_USR_BIN _BB_DIR_BIN
+#define _BB_DIR_USR_SBIN _BB_DIR_SBIN
 
 #ifdef CONFIG_TEST
 	APPLET_NOUSAGE("[", test_main, _BB_DIR_USR_BIN, _BB_SUID_NEVER)
