diff -Naur dvbtext.orig/dvbtext.c dvbtext-0.1/dvbtext.c
--- dvbtext.orig/dvbtext.c	2001-02-07 11:05:46.000000000 +0100
+++ dvbtext-0.1/dvbtext.c	2004-02-27 13:20:33.000000000 +0100
@@ -30,11 +30,16 @@
 */
 
 #include <stdio.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <sys/ioctl.h>
 #include <fcntl.h>
 #include <unistd.h>
+#ifdef NEWSTRUCT
+#include <linux/dvb/dmx.h>
+#else
 #include <ost/dmx.h>
+#endif
 #include "tables.h"
 
 #define VERSION "0.1"
@@ -128,6 +133,23 @@
   }
 }
 
+#ifdef NEWSTRUCT
+void set_tt_filt(int fd,uint16_t tt_pid)
+{
+	struct dmx_pes_filter_params pesFilterParams;
+
+	pesFilterParams.pid     = tt_pid;
+	pesFilterParams.input   = DMX_IN_FRONTEND;
+	pesFilterParams.output  = DMX_OUT_TS_TAP;
+        pesFilterParams.pes_type = DMX_PES_OTHER;
+	pesFilterParams.flags   = DMX_IMMEDIATE_START;
+
+	if (ioctl(fd, DMX_SET_PES_FILTER, &pesFilterParams) < 0)  {
+                fprintf(stderr,"FILTER %i: ",tt_pid);
+		perror("DMX SET PES FILTER");
+        }
+}
+#else
 void set_tt_filt(int fd,uint16_t tt_pid)
 {
 	struct dmxPesFilterParams pesFilterParams;
@@ -135,14 +157,15 @@
 	pesFilterParams.pid     = tt_pid;
 	pesFilterParams.input   = DMX_IN_FRONTEND;
 	pesFilterParams.output  = DMX_OUT_TS_TAP;
-        pesFilterParams.pesType = DMX_PES_TELETEXT;
-	pesFilterParams.flags   = DMX_IMMEDIATE_START|DMX_ONESHOT;
+        pesFilterParams.pesType = DMX_PES_OTHER;
+	pesFilterParams.flags   = DMX_IMMEDIATE_START;
 
 	if (ioctl(fd, DMX_SET_PES_FILTER, &pesFilterParams) < 0)  {
                 fprintf(stderr,"FILTER %i: ",tt_pid);
 		perror("DMX SET PES FILTER");
         }
 }
+#endif
 
 int main(int argc, char **argv)
 {
