diff -urN vsftpd-1.0.1.orig/Makefile vsftpd-1.0.1/Makefile
--- vsftpd-1.0.1.orig/Makefile	Mon Oct  8 23:35:27 2001
+++ vsftpd-1.0.1/Makefile	Fri Mar  1 23:14:33 2002
@@ -1,10 +1,9 @@
 # Makefile for systems with GNU tools
-CC 	=	gcc
 INSTALL	=	install
 IFLAGS  = -idirafter dummyinc
-CFLAGS	=	-O2 -Wall -W -Wshadow #-pedantic -Werror -Wconversion
+CFLAGS	=	-Os -Wall -W -mcpu=823 -mmultiple -mstring -meabi -msoft-float
 
-LIBS	=	`./vsf_findlibs.sh`
+LIBS	=	-lcrypt
 LINK	=	-Wl,-s
 
 OBJS	=	main.o utility.o prelogin.o ftpcmdio.o postlogin.o privsock.o \
diff -urN vsftpd-1.0.1.orig/sysdeputil.c vsftpd-1.0.1/sysdeputil.c
--- vsftpd-1.0.1.orig/sysdeputil.c	Tue Nov 13 22:12:54 2001
+++ vsftpd-1.0.1/sysdeputil.c	Fri Mar  1 23:01:53 2002
@@ -39,10 +39,10 @@
 #undef VSF_SYSDEP_HAVE_HPUX_SETPROCTITLE
 #undef VSF_SYSDEP_HAVE_MAP_ANON
 #undef VSF_SYSDEP_NEED_OLD_FD_PASSING
-#define VSF_SYSDEP_HAVE_PAM
+#undef VSF_SYSDEP_HAVE_PAM
 #define VSF_SYSDEP_HAVE_SHADOW
 #define VSF_SYSDEP_HAVE_USERSHELL
-#define VSF_SYSDEP_HAVE_LIBCAP
+#undef VSF_SYSDEP_HAVE_LIBCAP
 
 /* BEGIN config */
 #if defined(__linux__) && !defined(__ia64__) && !defined(__s390__)
diff -urN vsftpd-1.0.1.orig/vsftpd-dbox2.conf vsftpd-1.0.1/vsftpd-dbox2.conf
--- vsftpd-1.0.1.orig/vsftpd-dbox2.conf	Thu Jan  1 01:00:00 1970
+++ vsftpd-1.0.1/vsftpd-dbox2.conf	Fri Mar  1 23:13:43 2002
@@ -0,0 +1,12 @@
+# config file /etc/vsftpd.conf
+
+anonymous_enable=NO
+local_enable=YES
+write_enable=YES
+local_umask=022
+dirmessage_enable=YES
+xferlog_enable=NO
+connect_from_port_20=YES
+secure_chroot_dir=/share/empty
+ftpd_banner=Willkomen auf Ihrer d-box 2.
+
