diff -urN gcc-3.0.3.orig/libstdc++-v3/src/Makefile.am gcc-3.0.3/libstdc++-v3/src/Makefile.am
--- gcc-3.0.3.orig/libstdc++-v3/src/Makefile.am	Mon Sep 10 22:11:06 2001
+++ gcc-3.0.3/libstdc++-v3/src/Makefile.am	Sat Jan 12 13:23:31 2002
@@ -137,3 +137,7 @@
 # compilation driver.
 CXXLINK = $(LIBTOOL) --tag CXX --mode=link $(CXX) \
 	  @OPT_LDFLAGS@ @SECTION_LDFLAGS@ $(AM_CXXFLAGS) $(LDFLAGS) -o $@
+
+install-exec-local:
+	$(AR) cru libstdc++_pic.a .libs/*.o ../libsupc++/*.o
+
--- gcc-3.1/libstdc++-v3/src/Makefile.in.orig	Tue Apr  2 04:47:04 2002
+++ gcc-3.1/libstdc++-v3/src/Makefile.in	Thu May 16 22:52:02 2002
@@ -434,7 +434,7 @@
 installcheck: installcheck-am
 install-info-am: 
 install-info: install-info-am
-install-exec-am: install-toolexeclibLTLIBRARIES
+install-exec-am: install-toolexeclibLTLIBRARIES install-exec-local
 install-exec: install-exec-am
 
 install-data-am:
@@ -498,15 +498,20 @@
 clean-libtool maintainer-clean-libtool tags mostlyclean-tags \
 distclean-tags clean-tags maintainer-clean-tags distdir info-am info \
 dvi-am dvi check check-am installcheck-am installcheck install-info-am \
-install-info install-exec-am install-exec install-data-am install-data \
-install-am install uninstall-am uninstall all-redirect all-am all \
-installdirs mostlyclean-generic distclean-generic clean-generic \
-maintainer-clean-generic clean mostlyclean distclean maintainer-clean
+install-info install-exec-local install-exec-am install-exec \
+install-data-am install-data install-am install uninstall-am uninstall \
+all-redirect all-am all installdirs mostlyclean-generic \
+distclean-generic clean-generic maintainer-clean-generic clean \
+mostlyclean distclean maintainer-clean
 
 strstream.lo: strstream.cc
 	$(LTCXXCOMPILE) -I$(GLIBCPP_INCLUDE_DIR)/backward -Wno-deprecated -c $<
 strstream.o: strstream.cc
 	$(CXXCOMPILE) -I$(GLIBCPP_INCLUDE_DIR)/backward -Wno-deprecated -c $<
+
+
+install-exec-local:
+	$(AR) cru libstdc++_pic.a .libs/*.o ../libsupc++/*.o
 
 # Use special rules for the concept-checking instantiations so that all
 # the generated template functions are also instantiated.  Force the checks
