--- gdb-5.2/bfd/doc/Makefile.in.old	Mon Feb 11 06:01:57 2002
+++ gdb-5.2/bfd/doc/Makefile.in	Tue May 21 14:56:57 2002
@@ -460,7 +460,7 @@
 	$(CC_FOR_BUILD) -o $(MKDOC) chew.o $(LOADLIBES) $(LDFLAGS)
 
 chew.o: chew.c
-	$(CC_FOR_BUILD) -c -I.. -I$(srcdir)/.. -I$(srcdir)/../../include -I$(srcdir)/../../intl -I../../intl $(H_CFLAGS) $(CFLAGS) $(srcdir)/chew.c
+	$(CC_FOR_BUILD) -c -I.. -I$(srcdir)/.. -I$(srcdir)/../../include -I$(srcdir)/../../intl -I../../intl $(H_CFLAGS) $(srcdir)/chew.c
 
 protos: libbfd.h libcoff.h bfd.h
 
--- gdb+dejagnu-20021116/configure.in	2002-11-16 02:55:14.000000000 +0100
+++ gdb+dejagnu-20021116new/configure.in	2002-11-16 23:52:59.000000000 +0100
@@ -238,6 +238,9 @@
   powerpc-*-beos*)
     noconfigdirs="$noconfigdirs tk itcl tix libgui gdb dejagnu readline"
     ;;
+ powerpc-tuxbox-linux-gnu)
+    noconfigdirs="tk tcl expect dejagnu"
+    ;;
 esac
 
 # Save it here so that, even in case of --enable-libgcj, if the Java
