diff -uNr ../gdbm-1.8.3.org/Makefile.in ./Makefile.in
--- ../Makefile.in	2002-10-08 18:09:12.000000000 +0200
+++ ../Makefile.in	2008-10-06 14:57:49.000000000 +0200
@@ -14,10 +14,6 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_DATA = @INSTALL_DATA@
 
-# File ownership and group
-BINOWN = bin
-BINGRP = bin
-
 MAKEINFO = makeinfo
 TEXI2DVI = texi2dvi
 
@@ -131,11 +127,11 @@
 		$(INSTALL_ROOT)$(includedir) $(INSTALL_ROOT)$(man3dir) \
 		$(INSTALL_ROOT)$(infodir)
 	$(LIBTOOL) $(INSTALL) -c libgdbm.la $(INSTALL_ROOT)$(libdir)/libgdbm.la
-	$(INSTALL_DATA) -o $(BINOWN) -g $(BINGRP) gdbm.h \
+	$(INSTALL_DATA) gdbm.h \
 		$(INSTALL_ROOT)$(includedir)/gdbm.h
-	$(INSTALL_DATA) -o $(BINOWN) -g $(BINGRP) $(srcdir)/gdbm.3 \
+	$(INSTALL_DATA) $(srcdir)/gdbm.3 \
 		$(INSTALL_ROOT)$(man3dir)/gdbm.3
-	$(INSTALL_DATA) -o $(BINOWN) -g $(BINGRP) $(srcdir)/gdbm.info \
+	$(INSTALL_DATA) $(srcdir)/gdbm.info \
 		$(INSTALL_ROOT)$(infodir)/gdbm.info
 
 install-compat:
@@ -143,9 +139,9 @@
 		$(INSTALL_ROOT)$(includedir)
 	$(LIBTOOL) $(INSTALL) -c libgdbm_compat.la \
 		$(INSTALL_ROOT)$(libdir)/libgdbm_compat.la
-	$(INSTALL_DATA) -o $(BINOWN) -g $(BINGRP) $(srcdir)/dbm.h \
+	$(INSTALL_DATA) $(srcdir)/dbm.h \
 		$(INSTALL_ROOT)$(includedir)/dbm.h
-	$(INSTALL_DATA) -o $(BINOWN) -g $(BINGRP) $(srcdir)/ndbm.h \
+	$(INSTALL_DATA) $(srcdir)/ndbm.h \
 		$(INSTALL_ROOT)$(includedir)/ndbm.h
 
 #libgdbm.a: $(OBJS) gdbm.h
diff -uNr gdbm-1.8.3.org/configure gdbm-1.8.3/configure
--- gdbm-1.8.3.org/configure	2002-10-15 23:53:26.000000000 +0200
+++ gdbm-1.8.3/configure	2008-10-09 21:58:50.000000000 +0200
@@ -2919,7 +2919,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   case $host_cpu in
   alpha* | hppa* | i*86 | powerpc* | sparc* | ia64* )
     lt_cv_deplibs_check_method=pass_all ;;
@@ -5532,7 +5532,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
