diff -urN openssl-0.9.7b.orig/Configure openssl-0.9.7b/Configure
--- openssl-0.9.7b.orig/Configure	2003-04-10 07:46:55.000000000 +0200
+++ openssl-0.9.7b/Configure	2003-04-11 17:56:00.000000000 +0200
@@ -556,6 +556,9 @@
 ##### OS/2 EMX
 "OS2-EMX", "gcc::::::::",
 
+##### Tuxbox
+"tuxbox",    "powerpc-tuxbox-linux-gnu-gcc:-DB_ENDIAN -DTERMIO -O3 -fomit-frame-pointer -Wall -mmultiple -mstring -meabi -msoft-float::-D_REENTRANT::-ldl:BN_LLONG RC4_CHAR RC4_CHUNK DES_RISC1 DES_UNROLL::::::::::dlfcn:linux-shared:-fPIC::.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
+
 ##### VxWorks for various targets
 "vxworks-ppc405","ccppc:-g -msoft-float -mlongcall -DCPU=PPC405 -I\$(WIND_BASE)/target/h:::VXWORKS:-r:::::",
 "vxworks-ppc750","ccppc:-ansi -nostdinc -DPPC750 -D_REENTRANT -fvolatile -fno-builtin -fno-for-scope -fsigned-char -Wall -msoft-float -mlongcall -DCPU=PPC604 -I\$(WIND_BASE)/target/h \$(DEBUG_FLAG):::VXWORKS:-r:::::",
diff -urN openssl-0.9.7b.orig/Makefile.org openssl-0.9.7b/Makefile.org
--- openssl-0.9.7b.orig/Makefile.org	2003-04-08 13:54:32.000000000 +0200
+++ openssl-0.9.7b/Makefile.org	2003-04-11 17:56:00.000000000 +0200
@@ -173,7 +173,7 @@
 # we might set SHLIB_MARK to '$(SHARED_LIBS)'.
 SHLIB_MARK=
 
-DIRS=   crypto ssl $(SHLIB_MARK) apps test tools
+DIRS=   crypto ssl
 SHLIBDIRS= crypto ssl
 
 # dirs in crypto to build
@@ -877,4 +877,40 @@
 			 done); \
 	done
 
+install_cdk:
+	@mkdir -p \
+		$(INSTALL_PREFIX)$(INSTALLTOP)/lib/pkgconfig \
+		$(INSTALL_PREFIX)$(INSTALLTOP)/include/openssl \
+		$(INSTALL_PREFIX)$(OPENSSLDIR)/lib
+	@for i in $(EXHEADER) ;\
+	do \
+	(cp $$i $(INSTALL_PREFIX)$(INSTALLTOP)/include/openssl/$$i; \
+	chmod 644 $(INSTALL_PREFIX)$(INSTALLTOP)/include/openssl/$$i ); \
+	done;
+	@for i in $(SHLIBDIRS) ;\
+	do \
+	if [ -d "$$i" ]; then \
+		(cd $$i; echo "installing $$i..."; \
+		$(MAKE) INSTALL_PREFIX='${INSTALL_PREFIX}' INSTALLTOP='${INSTALLTOP}' install ); \
+	fi; \
+	done	
+	@if [ -n "$(SHARED_LIBS)" ]; then \
+		tmp="$(SHARED_LIBS)"; \
+		for i in $${tmp:-x}; \
+		do \
+			if [ -f "$$i" -o -f "$$i.a" ]; then \
+				echo installing $$i; \
+				cp $$i $(INSTALL_PREFIX)$(INSTALLTOP)/lib/$$i.new; \
+				chmod 555 $(INSTALL_PREFIX)$(INSTALLTOP)/lib/$$i.new; \
+				mv -f $(INSTALL_PREFIX)$(INSTALLTOP)/lib/$$i.new $(INSTALL_PREFIX)$(INSTALLTOP)/lib/$$i; \
+			fi; \
+		done; \
+		here="`pwd`"; \
+		cd $(INSTALL_PREFIX)$(INSTALLTOP)/lib; \
+		set $(MAKE); \
+		$$1 -f $$here/Makefile link-shared; \
+	fi
+	cp openssl.pc $(INSTALL_PREFIX)$(INSTALLTOP)/lib/pkgconfig
+
+
 # DO NOT DELETE THIS LINE -- make depend depends on it.
