--- libdvb-0.5.0/Makefile	2003-08-29 16:26:19.000000000 +0200
+++ libdvb-0.5.0new/Makefile	2003-09-08 13:12:45.000000000 +0200
@@ -18,7 +18,6 @@
 	install -m 644 $(DVB_LIBS) $(PREFIX)/lib/
 	[ ! -d "$(PREFIX)/include" ] && mkdir -p $(PREFIX)/include
 	install -m 644 $(HEADERS) $(PREFIX)/include/
-	make -C dvb-mpegtools install
 
 uninstall: 
 	for i in $(DVB_LIBS); do rm -f $(PREFIX)/lib/$$i;done
