diff -Naur libdvb-0.5.5.orig/Makefile libdvb-0.5.5/Makefile
--- libdvb-0.5.5.orig/Makefile	2004-09-15 20:56:21.000000000 +0200
+++ libdvb-0.5.5/Makefile	2005-10-18 23:01:55.000000000 +0200
@@ -18,7 +18,6 @@
 	install -m 644 $(DVB_LIBS) $(DESTDIR)$(PREFIX)/lib/
 	mkdir -p $(DESTDIR)$(PREFIX)/include
 	install -m 644 $(HEADERS) $(DESTDIR)$(PREFIX)/include/
-	make -C dvb-mpegtools install
 
 uninstall: 
 	for i in $(DVB_LIBS); do rm -f $(DESTDIR)$(PREFIX)/lib/$$i;done
diff -Naur libdvb-0.5.5.orig/include/DVB.hh libdvb-0.5.5/include/DVB.hh
--- libdvb-0.5.5.orig/include/DVB.hh	2004-09-15 20:56:21.000000000 +0200
+++ libdvb-0.5.5/include/DVB.hh	2005-10-18 23:03:21.000000000 +0200
@@ -2,6 +2,7 @@
 #define _DVB_DEV_HH_
 
 extern "C" {
+#include <errno.h>
 #include <asm/errno.h>
 #include <fcntl.h>
 #include <netdb.h>
