diff -Naur freetype-2.1.9-orig/builds/unix/freetype-config.in freetype-2.1.9/builds/unix/freetype-config.in
--- freetype-2.1.9-orig/builds/unix/freetype-config.in	2004-04-17 20:58:43.000000000 +0200
+++ freetype-2.1.9/builds/unix/freetype-config.in	2005-01-03 20:18:49.004146232 +0100
@@ -106,11 +106,7 @@
 
 if test "$echo_cflags" = "yes" ; then
   cflags="-I$includedir/freetype2"
-  if test "$includedir" != "/usr/include" ; then
-    echo $cflags -I$includedir
-  else
     echo $cflags
-  fi
 fi
 
 if test "$echo_libs" = "yes" ; then
