diff -Naur freetype-2.1.7.ori/builds/unix/freetype-config.in freetype-2.1.7/builds/unix/freetype-config.in
--- freetype-2.1.7.ori/builds/unix/freetype-config.in	2003-04-24 07:45:59.000000000 +0200
+++ freetype-2.1.7/builds/unix/freetype-config.in	2004-04-11 20:04:34.000000000 +0200
@@ -80,11 +80,7 @@
 
 if test "$echo_cflags" = "yes" ; then
  cflags="-I@includedir@/freetype2"
- if test "@includedir@" != "/usr/include" ; then
-  echo $cflags -I@includedir@
- else
   echo $cflags
- fi
 fi
 
 if test "$echo_libs" = "yes" ; then
