--- old/configure.ac	2006-07-06 19:06:25.000000000 +0200
+++ new/configure.ac	2006-07-06 19:08:36.000000000 +0200
@@ -52,7 +52,9 @@
 AC_STRUCT_TM
 
 # Checks for library functions.
-AC_FUNC_MALLOC
+dnl AC_FUNC_MALLOC always returns failure when cross compiling, therefore
+dnl we nuke the test. /Barf 2006-07-06
+dnl AC_FUNC_MALLOC
 AC_FUNC_STRTOD
 AC_CHECK_FUNCS([memset], , AC_ERROR([memset not found in libc]))
 AC_CHECK_FUNCS([pow], , AC_CHECK_LIB(m, pow, , AC_ERROR([cannot find pow])) )
