Index: b/kernel/timer.c
===================================================================
--- a/kernel/timer.c
+++ b/kernel/timer.c
@@ -33,11 +33,14 @@ long tick = (1000000 + HZ/2) / HZ;	/* ti
 
 /* The current time */
 struct timeval xtime __attribute__ ((aligned (16)));
 
 /* Don't completely fail for HZ > 500.  */
-int tickadj = 500/HZ ? : 1;		/* microsecs */
+/* due to the dbox inexact timer, we allow more than
+   500 microsecs per tick Orignal would correct up to
+   0.5%, change it to 2%... */
+int tickadj = 2000/HZ ? : 1;		/* microsecs */
 
 DECLARE_TASK_QUEUE(tq_timer);
 DECLARE_TASK_QUEUE(tq_immediate);
 
 /*
