diff -Naur linux-2.4.32/arch/ppc/kernel/m8xx_setup.c linux-2.4.32-ide2/arch/ppc/kernel/m8xx_setup.c
--- linux-2.4.32/arch/ppc/kernel/m8xx_setup.c	2004-02-18 14:36:30.000000000 +0100
+++ linux-2.4.32-ide2/arch/ppc/kernel/m8xx_setup.c	2006-06-28 10:03:06.000000000 +0200
@@ -410,7 +410,9 @@
 	ppc_md.kbd_init_hw		= NULL;
 	ppc_md.ppc_kbd_sysrq_xlate	= NULL;
 
+#if defined(CONFIG_BLK_DEV_MPC8xx_IDE)
 #if defined(CONFIG_BLK_DEV_IDE) || defined(CONFIG_BLK_DEV_IDE_MODULE)
 	m8xx_ide_init();
 #endif
+#endif
 }
