diff -ur linux-2.4.27-dbox/arch/ppc/platforms/dbox2.h linux-2.4.27/arch/ppc/platforms/dbox2.h
--- linux-2.4.27-dbox/arch/ppc/platforms/dbox2.h	Fri Nov 28 19:26:19 2003
+++ linux/arch/ppc/platforms/dbox2.h	Fri Feb 20 21:00:19 2004
@@ -1,6 +1,6 @@
 /*
  * D-BOX2 board specific definitions
- * 
+ *
  * Copyright (c) 2001-2002 Florian Schirmer (jolt@tuxbox.org)
  */
 
@@ -8,7 +8,7 @@
 #define __MACH_DBOX2_H
 
 #include <linux/config.h>
- 
+
 #include <asm/ppcboot.h>
 
 #define	DBOX2_IMMR_BASE	0xFF000000	/* phys. addr of IMMR */
@@ -16,6 +16,83 @@
 
 #define	IMAP_ADDR	DBOX2_IMMR_BASE	/* physical base address of IMMR area */
 #define IMAP_SIZE	DBOX2_IMAP_SIZE	/* mapped size of IMMR area */
+
+
+/*-----------------------------------------------------------------------
+ * added 11.09.2003 07:14 by RR4711 for IDE Interface support
+ *-----------------------------------------------------------------------
+*/
+
+#if (defined(CONFIG_PCMCIA_M8XX)    || defined(CONFIG_PCMCIA_M8XX_MODULE) ) && \
+    (defined(CONFIG_IDE_8xx_PCCARD) || defined(CONFIG_IDE_8xx_DIRECT) )
+#error "Do NOT defined CONFIG_PCMCIA_M8XX _and_ CONFIG_IDE_8xx_XXX !!!"
+#endif
+
+
+/*if defined(CONFIG_PCMCIA_M8XX) || defined(CONFIG_PCMCIA_M8XX_MODULE)*/
+#ifdef CONFIG_HOTPLUG
+/* define IO_BASE for PCMCIA */
+#define _IO_BASE	0x80000000
+#define _IO_BASE_SIZE	(64<<10)
+/*#define _IO_BASE_SIZE	0x1000		??? */
+
+#ifdef CONFIG_IDE
+#define ide_request_irq(irq,hand,flg,dev,id)    \
+        request_8xxirq((irq),(hand),(flg),(dev),(id))
+#endif
+#endif
+
+#define PCMCIA_MEM_SIZE		( 64 << 20 )
+
+#define MAX_HWIFS			1	/* overwrite default in include/asm-ppc/ide.h	*/
+
+/*
+ * Definitions for IDE0 Interface
+ */
+
+#if defined(CONFIG_IDE_8xx_PCCARD)
+# define IDE0_BASE_OFFSET		0
+# define IDE0_DATA_REG_OFFSET		(PCMCIA_MEM_SIZE + 0x320)
+# define IDE0_ERROR_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 1)
+# define IDE0_NSECTOR_REG_OFFSET	(2 * PCMCIA_MEM_SIZE + 0x320 + 2)
+# define IDE0_SECTOR_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 3)
+# define IDE0_LCYL_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 4)
+# define IDE0_HCYL_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 5)
+# define IDE0_SELECT_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 6)
+# define IDE0_STATUS_REG_OFFSET		(2 * PCMCIA_MEM_SIZE + 0x320 + 7)
+# define IDE0_CONTROL_REG_OFFSET	0x0106
+# define IDE0_IRQ_REG_OFFSET		0x000A	/* not used			*/
+
+#elif defined(CONFIG_IDE_EXT_DIRECT)
+
+#define CFG_ATA_BASE_ADDR		0x02000000
+
+#define IDE0_BASE_OFFSET		0x0000
+#define IDE0_DATA_REG_OFFSET		0x0000
+#define IDE0_ERROR_REG_OFFSET		0x0004
+#define IDE0_NSECTOR_REG_OFFSET		0x0008
+#define IDE0_SECTOR_REG_OFFSET		0x000c
+#define IDE0_LCYL_REG_OFFSET		0x0010
+#define IDE0_HCYL_REG_OFFSET		0x0014
+#define IDE0_SELECT_REG_OFFSET		0x0018
+#define IDE0_STATUS_REG_OFFSET		0x001c
+
+#define IDE0_CONTROL_REG_OFFSET		0x0038
+#define IDE0_IRQ_REG_OFFSET		0x0008
+#endif
+
+/*
+ * Interrupts
+ */
+
+#define IDE0_INTERRUPT		SIU_IRQ6	/*  external interrupt	*/
+
+#define DAB4K_IDE_TIMING_STRING	"(OR_CSNT | OR_ACS_10 | OR_BI | OR_SCY_1_CLK | OR_TRLX | OR_EHTR)"
+
+#define CFG_OR			memc_or2
+#define CFG_BR			memc_br2
+#define CFG_OR_VALUE		0xfe000966
+#define CFG_BR_VALUE		0x02000001
 
 /* We don't use the 8259.
 */
diff -ur linux-2.4.27-dbox/drivers/ide/ide-probe.c linux-2.4.27/drivers/ide/ide-probe.c
--- linux-2.4.27-dbox/drivers/ide/ide-probe.c	Fri Nov 28 19:26:20 2003
+++ linux/drivers/ide/ide-probe.c	Fri Feb 20 21:00:19 2004
@@ -175,6 +175,7 @@
 	/* we depend on this a lot! */
 	id->model[sizeof(id->model)-1] = '\0';
 	printk("%s: %s, ", drive->name, id->model);
+
 	drive->present = 1;
 	drive->dead = 0;
 
@@ -261,6 +262,7 @@
 	}
 	drive->media = ide_disk;
 	printk("%s DISK drive\n", (drive->is_flash) ? "CFA" : "ATA" );
+
 	QUIRK_LIST(drive);
 	return;
 
diff -ur linux-2.4.27-dbox/drivers/ide/ppc/mpc8xx.c linux-2.4.27/drivers/ide/ppc/mpc8xx.c
--- linux-2.4.27-dbox/drivers/ide/ppc/mpc8xx.c	Fri Jun 13 16:51:34 2003
+++ linux/drivers/ide/ppc/mpc8xx.c	Fri Feb 20 21:00:19 2004
@@ -31,6 +31,8 @@
 #include <linux/ioport.h>
 #include <linux/ide.h>
 #include <linux/bootmem.h>
+#include <linux/delay.h>
+
 
 #include <asm/mpc8xx.h>
 #include <asm/mmu.h>
@@ -44,6 +46,7 @@
 #include <asm/irq.h>
 
 #include "ide_modes.h"
+
 static int identify  (volatile u8 *p);
 static void print_fixed (volatile u8 *p);
 static void print_funcid (int func);
@@ -356,17 +359,23 @@
 	static unsigned long ide_base = 0;
 	unsigned long base;
 
+	volatile memctl8xx_t *memctl = &(((volatile immap_t *) IMAP_ADDR)->im_memctl);
+
 	*p = 0;
 	if (irq)
 		*irq = 0;
 
 	if (!ide_base) {
 
-		/* TODO:
-		 * - add code to read ORx, BRx
-		 */
+		memctl->CFG_OR = CFG_OR_VALUE;
+		memctl->CFG_BR = CFG_BR_VALUE;
+		
+		printk ("MEMCTRL_OR : %08x\n", CFG_OR_VALUE);
+		printk ("MEMCTRL_BR : %08x\n", CFG_BR_VALUE);
+
+
 		ide_phy_base = CFG_ATA_BASE_ADDR;
-		ide_phy_end  = CFG_ATA_BASE_ADDR + 0x200;
+		ide_phy_end  = CFG_ATA_BASE_ADDR + 0x4000;
 
 		printk ("IDE phys mem : %08x...%08x (size %08x)\n",
 			ide_phy_base, ide_phy_end,
@@ -379,6 +388,14 @@
 		printk ("IDE virt base: %08lx\n", ide_base);
 #endif
 	}
+//	printk ("Trying to set LEDs @%08lx using 8 bit wide access...\n",ide_base+(0x800));
+//	outb(0x55,ide_base+(0x800));
+	
+//	printk ("Trying to set LEDs @%08lx using 16 bit wide access...\n",ide_base+(0x800));
+//	outw(0x00AA,ide_base+(0x800));
+	
+//	udelay(1000000);
+	
 
 	if (data_port >= MAX_HWIFS)
 		return;
@@ -533,6 +550,7 @@
 	 * The interrupt is handled somewhere else.	-- Steven
 	 */
 #endif /* CONFIG_IDE_8xx_PCCARD */
+return;
 }
 
 
@@ -855,3 +873,5 @@
 	ppc_ide_md.default_io_base      = m8xx_ide_default_io_base;
 	ppc_ide_md.ide_init_hwif        = m8xx_ide_init_hwif_ports;
 }
+
+#undef DEBUG
diff -ur linux-2.4.27-dbox/drivers/mtd/maps/dbox2-flash.c linux-2.4.27/drivers/mtd/maps/dbox2-flash.c
--- linux-2.4.27-dbox/drivers/mtd/maps/dbox2-flash.c	Sat Feb 21 14:39:09 2004
+++ linux/drivers/mtd/maps/dbox2-flash.c	Fri Feb 20 21:00:19 2004
@@ -33,13 +33,13 @@
 	},
 	{
 	.name		= "root (cramfs)",	
-	.size		= 7040 * 1024, 
+	.size		= 7168 * 1024, 
 	.offset		= MTDPART_OFS_APPEND, 
 	.mask_flags	= 0
 	},
 	{
 	.name		= "var (jffs2)",
-	.size		= 896 * 1024, 
+	.size		= 768 * 1024, 
 	.offset		= MTDPART_OFS_APPEND, 
 	.mask_flags	= 0
 	},
diff -ur linux-2.4.27-dbox/include/asm-ppc/ide.h linux-2.4.27/include/asm-ppc/ide.h
--- linux-2.4.27-dbox/include/asm-ppc/ide.h	Fri Jun 13 16:51:38 2003
+++ linux/include/asm-ppc/ide.h	Fri Feb 20 21:03:18 2004
@@ -17,7 +17,7 @@
 #include <asm/mpc8xx.h>
 
 #ifndef MAX_HWIFS
-#define MAX_HWIFS	8
+#define MAX_HWIFS	1
 #endif
 
 #include <asm/hdreg.h>
@@ -108,7 +108,7 @@
  * The following are not needed for the non-m68k ports
  * unless direct IDE on 8xx
  */
-#if (defined CONFIG_APUS || defined CONFIG_BLK_DEV_MPC8xx_IDE )
+#if (defined CONFIG_APUS)
 #define IDE_ARCH_ACK_INTR 1
 #endif
 
