diff -Naur linux-2.4.22-pre6/arch/ppc/kernel/irq.c linux-2.4.22-pre6-dbox2/arch/ppc/kernel/irq.c
--- linux-2.4.22-pre6/arch/ppc/kernel/irq.c	2003-06-13 16:51:31.000000000 +0200
+++ linux-2.4.22-pre6-dbox2/arch/ppc/kernel/irq.c	2003-07-15 09:16:34.000000000 +0200
@@ -507,6 +507,14 @@
 		else if (irq_desc[irq].handler->enable)
 			irq_desc[irq].handler->enable(irq);
 	}
+	
+#ifdef CONFIG_DBOX2
+	if (action) {
+	    if (action->flags & SA_ONESHOT)
+		disable_irq_nosync(irq);
+	}
+#endif
+
 	spin_unlock(&desc->lock);
 }
 
