diff -Naur linux-2.4.33.3/arch/ppc/kernel/m8xx_setup.c linux-2.4.33.3-ide/arch/ppc/kernel/m8xx_setup.c
--- linux-2.4.33.3/arch/ppc/kernel/m8xx_setup.c	2006-08-31 19:03:20.000000000 +0200
+++ linux-2.4.33.3-ide/arch/ppc/kernel/m8xx_setup.c	2006-09-19 00:05:28.000000000 +0200
@@ -410,7 +410,9 @@
 	ppc_md.kbd_init_hw		= NULL;
 	ppc_md.ppc_kbd_sysrq_xlate	= NULL;
 
+#if defined(CONFIG_BLK_DEV_MPC8xx_IDE)
 #if defined(CONFIG_BLK_DEV_IDE) || defined(CONFIG_BLK_DEV_IDE_MODULE)
 	m8xx_ide_init();
 #endif
+#endif
 }
diff -Naur linux-2.4.33.3/include/asm-ppc/pgtable.h linux-2.4.33.3-ide/include/asm-ppc/pgtable.h
--- linux-2.4.33.3/include/asm-ppc/pgtable.h	2006-08-31 19:03:20.000000000 +0200
+++ linux-2.4.33.3-ide/include/asm-ppc/pgtable.h	2006-09-16 10:57:48.000000000 +0200
@@ -632,8 +632,8 @@
 #define SWP_TYPE(entry)			((entry).val & 0x3f)
 #define SWP_OFFSET(entry)		((entry).val >> 6)
 #define SWP_ENTRY(type, offset)		((swp_entry_t) { (type) | ((offset) << 6) })
-#define pte_to_swp_entry(pte)		((swp_entry_t) { pte_val(pte) >> 2 })
-#define swp_entry_to_pte(x)		((pte_t) { (x).val << 2 })
+#define pte_to_swp_entry(pte)		((swp_entry_t) { pte_val(pte) >> 8 })
+#define swp_entry_to_pte(x)		((pte_t) { (x).val << 8 })
 
 /* CONFIG_APUS */
 /* For virtual address to physical address conversion */
