--- linux-2.4.20/drivers/mtd/chips/cfi_probe.c	Fri Jul 18 18:11:16 2003
+++ linux-2.4.20-dbox2/drivers/mtd/chips/cfi_probe.c	Fri Oct  5 00:14:59 2001
@@ -123,14 +123,9 @@
 	int ofs_factor = cfi->interleave*cfi->device_type;
 	__u32 base = 0;
 	int num_erase_regions = cfi_read_query(map, base + (0x10 + 28)*ofs_factor);
-	int man_id = cfi_read_query(map, base + (0x00)*ofs_factor);
-	int dev_id = cfi_read_query(map, base + (0x01)*ofs_factor);
-
 	int i;
 
 #ifdef DEBUG_CFI
-	printk("Man.-ID  : %x\n", man_id);
-	printk("Device-Id: %x\n", dev_id);
 	printk("Number of erase regions: %d\n", num_erase_regions);
 #endif
 	if (!num_erase_regions)
@@ -169,13 +164,7 @@
 	for (i=0; i<cfi->cfiq->NumEraseRegions; i++) {
 		cfi->cfiq->EraseRegionInfo[i] = le32_to_cpu(cfi->cfiq->EraseRegionInfo[i]);
 		
-		if ((man_id==0x20) && (dev_id==0xBB) && (i==1)) // ST M28W320CB report wrong EraseRegionInfo for Region 2
-			cfi->cfiq->EraseRegionInfo[i] = 0x100003E;
-
 #ifdef DEBUG_CFI		
-
-		printk("EraseRegionInfo[%d], %x\n", i, cfi->cfiq->EraseRegionInfo[i]);
-
 		printk("  Erase Region #%d: BlockSize 0x%4.4X bytes, %d blocks\n",
 		       i, (cfi->cfiq->EraseRegionInfo[i] >> 8) & ~0xff, 
 		       (cfi->cfiq->EraseRegionInfo[i] & 0xffff) + 1);
