diff -rup linux-2.6.23~/init/do_mounts.c linux-2.6.23/init/do_mounts.c
--- linux-2.6.23~/init/do_mounts.c	2007-10-09 22:31:38.000000000 +0200
+++ linux-2.6.23/init/do_mounts.c	2007-11-18 00:36:10.000000000 +0100
@@ -291,6 +291,37 @@ static int __init do_mount_root(char *na
 	       current->fs->pwdmnt->mnt_sb->s_type->name,
 	       current->fs->pwdmnt->mnt_sb->s_flags & MS_RDONLY ? 
 	       " readonly" : "");
+
+	err = sys_mount("dev", "dev", "tmpfs", 0, NULL);
+	if (err < 0) {
+		printk("could not mount /dev tmpfs! bailing out\n");
+		return err;
+	}
+
+	err = sys_mknod((const char __user *) "dev/console",
+			S_IFCHR | S_IRUSR | S_IWUSR,
+			new_encode_dev(MKDEV(5, 1)));
+	if (err < 0) {
+		printk("could not create dev/console\n");
+		return err;
+	}
+
+	err = sys_mknod((const char __user *) "dev/null",
+			S_IFCHR | S_IRUSR | S_IWUSR,
+			new_encode_dev(MKDEV(1, 3)));
+	if (err < 0) {
+		printk("could not create dev/null\n");
+		return err;
+	}
+
+	err = sys_mknod((const char __user *) "dev/ttyCPM0",
+			S_IFCHR | S_IRUSR | S_IWUSR,
+			new_encode_dev(MKDEV(4, 64)));
+	if (err < 0) {
+		printk("could not create dev/ttyCPM0\n");
+		return err;
+	}
+
 	return 0;
 }
 
diff -rup linux-2.6.23~/kernel/printk.c linux-2.6.23/kernel/printk.c
--- linux-2.6.23~/kernel/printk.c	2007-10-09 22:31:38.000000000 +0200
+++ linux-2.6.23/kernel/printk.c	2007-11-18 13:08:29.000000000 +0100
@@ -682,6 +682,9 @@ static int __init console_setup(char *st
 	if (!strcmp(str, "ttyb"))
 		strcpy(buf, "ttyS1");
 #endif
+	if (!strcmp(str, "ttyS0"))
+		strcpy(buf, "ttyCPM0");
+
 	for (s = buf; *s; s++)
