diff -Naur ltrace-0.3.26.orig/Makefile.in ltrace-0.3.26/Makefile.in
--- ltrace-0.3.26.orig/Makefile.in	Sun Mar  3 22:00:43 2002
+++ ltrace-0.3.26/Makefile.in	Sat Aug 24 14:14:28 2002
@@ -48,7 +48,7 @@
 install:	ltrace
 		@INSTALL@ -d $(DESTDIR)$(bindir) $(DESTDIR)$(docdir) $(DESTDIR)$(mandir)/man1
 		@INSTALL@ -d $(DESTDIR)$(sysconfdir)
-		@INSTALL@ -s ltrace $(DESTDIR)$(bindir)
+		@INSTALL@ -m 755 ltrace $(DESTDIR)$(bindir)
 		@INSTALL@ -m 644 etc/ltrace.conf $(DESTDIR)$(sysconfdir)
 		@INSTALL@ -m 644 COPYING README TODO BUGS ChangeLog $(DESTDIR)$(docdir)
 		@INSTALL@ -m 644 ltrace.1 $(DESTDIR)$(mandir)/man1
diff -Naur ltrace-0.3.26.orig/sysdeps/linux-gnu/Makefile ltrace-0.3.26/sysdeps/linux-gnu/Makefile
--- ltrace-0.3.26.orig/sysdeps/linux-gnu/Makefile	Sun Mar  3 02:18:47 2002
+++ ltrace-0.3.26/sysdeps/linux-gnu/Makefile	Sat Aug 24 14:13:56 2002
@@ -1,4 +1,4 @@
-ARCH	:=	$(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ -e s/arm.*/arm/ -e s/sa110/arm/)
+ARCH	:=	ppc
 
 CFLAGS	+=	-I$(TOPDIR)/sysdeps/linux-gnu/$(ARCH)
 
