diff -urN mad-0.14.2b.orig/libid3tag/Makefile.am mad-0.14.2b/libid3tag/Makefile.am
--- mad-0.14.2b.orig/libid3tag/Makefile.am	2001-11-09 05:55:21.000000000 +0100
+++ mad-0.14.2b/libid3tag/Makefile.am	2001-11-09 05:55:21.000000000 +0100
@@ -52,6 +52,9 @@
 
 libid3tag_la_LDFLAGS =	-version-info $(version_info)
 
+pkgconfigdir =		$(libdir)/pkgconfig
+pkgconfig_DATA =	id3tag.pc
+
 BUILT_SOURCES =		frametype.c compat.c genre.dat
 
 $(srcdir)/frametype.c: $(srcdir)/frametype.gperf Makefile.am
diff -urN mad-0.14.2b.orig/libid3tag/Makefile.in mad-0.14.2b/libid3tag/Makefile.in
--- mad-0.14.2b.orig/libid3tag/Makefile.in	2001-11-09 08:02:58.000000000 +0100
+++ mad-0.14.2b/libid3tag/Makefile.in	2001-11-09 08:02:58.000000000 +0100
@@ -93,11 +93,14 @@
 
 libid3tag_la_LDFLAGS = -version-info $(version_info)
 
+pkgconfigdir = $(libdir)/pkgconfig
+pkgconfig_DATA = id3tag.pc
+
 BUILT_SOURCES = frametype.c compat.c genre.dat
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
 mkinstalldirs = $(SHELL) $(top_srcdir)/../mkinstalldirs
 CONFIG_HEADER = config.h
-CONFIG_CLEAN_FILES = 
+CONFIG_CLEAN_FILES =  id3tag.pc
 LTLIBRARIES =  $(lib_LTLIBRARIES)
 
 
@@ -118,10 +121,11 @@
 LTCOMPILE = $(LIBTOOL) --mode=compile $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
 CCLD = $(CC)
 LINK = $(LIBTOOL) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(LDFLAGS) -o $@
+DATA =  $(pkgconfig_DATA)
 HEADERS =  $(include_HEADERS)
 
 DIST_COMMON =  ./stamp-h.in Makefile.am Makefile.in acconfig.h \
-acinclude.m4 aclocal.m4 config.h.in configure configure.in
+acinclude.m4 aclocal.m4 config.h.in configure configure.in id3tag.pc.in
 
 
 DISTFILES = $(DIST_COMMON) $(SOURCES) $(HEADERS) $(TEXINFOS) $(EXTRA_DIST)
@@ -176,6 +180,8 @@
 	-rm -f config.h
 
 maintainer-clean-hdr:
+id3tag.pc: $(top_builddir)/config.status id3tag.pc.in
+	cd $(top_builddir) && CONFIG_FILES=$@ CONFIG_HEADERS= $(SHELL) ./config.status
 
 mostlyclean-libLTLIBRARIES:
 
@@ -243,6 +249,25 @@
 libid3tag.la: $(libid3tag_la_OBJECTS) $(libid3tag_la_DEPENDENCIES)
 	$(LINK) -rpath $(libdir) $(libid3tag_la_LDFLAGS) $(libid3tag_la_OBJECTS) $(libid3tag_la_LIBADD) $(LIBS)
 
+install-pkgconfigDATA: $(pkgconfig_DATA)
+	@$(NORMAL_INSTALL)
+	$(mkinstalldirs) $(DESTDIR)$(pkgconfigdir)
+	@list='$(pkgconfig_DATA)'; for p in $$list; do \
+	  if test -f $(srcdir)/$$p; then \
+	    echo " $(INSTALL_DATA) $(srcdir)/$$p $(DESTDIR)$(pkgconfigdir)/$$p"; \
+	    $(INSTALL_DATA) $(srcdir)/$$p $(DESTDIR)$(pkgconfigdir)/$$p; \
+	  else if test -f $$p; then \
+	    echo " $(INSTALL_DATA) $$p $(DESTDIR)$(pkgconfigdir)/$$p"; \
+	    $(INSTALL_DATA) $$p $(DESTDIR)$(pkgconfigdir)/$$p; \
+	  fi; fi; \
+	done
+
+uninstall-pkgconfigDATA:
+	@$(NORMAL_UNINSTALL)
+	list='$(pkgconfig_DATA)'; for p in $$list; do \
+	  rm -f $(DESTDIR)$(pkgconfigdir)/$$p; \
+	done
+
 install-includeHEADERS: $(include_HEADERS)
 	@$(NORMAL_INSTALL)
 	$(mkinstalldirs) $(DESTDIR)$(includedir)
@@ -350,20 +375,22 @@
 install-exec-am: install-libLTLIBRARIES
 install-exec: install-exec-am
 
-install-data-am: install-includeHEADERS
+install-data-am: install-pkgconfigDATA install-includeHEADERS
 install-data: install-data-am
 
 install-am: all-am
 	@$(MAKE) $(AM_MAKEFLAGS) install-exec-am install-data-am
 install: install-am
-uninstall-am: uninstall-libLTLIBRARIES uninstall-includeHEADERS
+uninstall-am: uninstall-libLTLIBRARIES uninstall-pkgconfigDATA \
+		uninstall-includeHEADERS
 uninstall: uninstall-am
-all-am: Makefile $(LTLIBRARIES) $(HEADERS) config.h
+all-am: Makefile $(LTLIBRARIES) $(DATA) $(HEADERS) config.h
 all-redirect: all-am
 install-strip:
 	$(MAKE) $(AM_MAKEFLAGS) AM_INSTALL_PROGRAM_FLAGS=-s install
 installdirs:
-	$(mkinstalldirs)  $(DESTDIR)$(libdir) $(DESTDIR)$(includedir)
+	$(mkinstalldirs)  $(DESTDIR)$(libdir) $(DESTDIR)$(pkgconfigdir) \
+		$(DESTDIR)$(includedir)
 
 
 mostlyclean-generic:
@@ -412,9 +439,11 @@
 uninstall-libLTLIBRARIES install-libLTLIBRARIES mostlyclean-compile \
 distclean-compile clean-compile maintainer-clean-compile \
 mostlyclean-libtool distclean-libtool clean-libtool \
-maintainer-clean-libtool uninstall-includeHEADERS \
-install-includeHEADERS tags mostlyclean-tags distclean-tags clean-tags \
-maintainer-clean-tags distdir info-am info dvi-am dvi check check-am \
+maintainer-clean-libtool uninstall-pkgconfigDATA install-pkgconfigDATA \
+uninstall-includeHEADERS install-includeHEADERS tags mostlyclean-tags \
+distclean-tags clean-tags maintainer-clean-tags distdir \
+mostlyclean-depend distclean-depend clean-depend \
+maintainer-clean-depend info-am info dvi-am dvi check check-am \
 installcheck-am installcheck all-recursive-am install-exec-am \
 install-exec install-data-am install-data install-am install \
 uninstall-am uninstall all-redirect all-am all installdirs \
diff -urN mad-0.14.2b.orig/libid3tag/configure mad-0.14.2b/libid3tag/configure
--- mad-0.14.2b.orig/libid3tag/configure	2001-11-02 10:52:25.000000000 +0100
+++ mad-0.14.2b/libid3tag/configure	2001-11-02 10:52:25.000000000 +0100
@@ -2301,7 +2301,7 @@
 
 cat >> $CONFIG_STATUS <<EOF
 
-CONFIG_FILES=\${CONFIG_FILES-"Makefile"}
+CONFIG_FILES=\${CONFIG_FILES-"Makefile id3tag.pc"}
 EOF
 cat >> $CONFIG_STATUS <<\EOF
 for ac_file in .. $CONFIG_FILES; do if test "x$ac_file" != x..; then
diff -urN mad-0.14.2b.orig/libid3tag/configure.in mad-0.14.2b/libid3tag/configure.in
--- mad-0.14.2b.orig/libid3tag/configure.in	2001-11-02 10:50:44.000000000 +0100
+++ mad-0.14.2b/libid3tag/configure.in	2001-11-02 10:50:44.000000000 +0100
@@ -156,4 +156,4 @@
 LTLIBOBJS=`echo "$LIBOBJS" | sed -e 's/\.o/.lo/g'`
 AC_SUBST(LTLIBOBJS)
 
-AC_OUTPUT(Makefile)
+AC_OUTPUT(Makefile id3tag.pc)
diff -urN mad-0.14.2b.orig/libid3tag/id3tag.pc.in mad-0.14.2b/libid3tag/id3tag.pc.in
--- mad-0.14.2b.orig/libid3tag/id3tag.pc.in	1970-01-01 01:00:00.000000000 +0100
+++ mad-0.14.2b/libid3tag/id3tag.pc.in	1970-01-01 01:00:00.000000000 +0100
@@ -0,0 +1,11 @@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+libdir=@libdir@
+includedir=@includedir@
+
+Name: id3tag
+Description: ID3 tag reading library
+Requires:
+Version: 0.14.2b
+Libs: -L${libdir} -lid3tag -lz
+Cflags: -I${includedir}
diff -urN mad-0.14.2b.orig/libmad/Makefile.am mad-0.14.2b/libmad/Makefile.am
--- mad-0.14.2b.orig/libmad/Makefile.am	2001-11-09 05:55:25.000000000 +0100
+++ mad-0.14.2b/libmad/Makefile.am	2001-11-09 05:55:25.000000000 +0100
@@ -44,6 +44,9 @@
 libmad_la_DEPENDENCIES =	@ASO_OBJS@
 libmad_la_LIBADD =		@ASO_OBJS@
 
+pkgconfigdir =			$(libdir)/pkgconfig
+pkgconfig_DATA =		mad.pc
+
 #CFLAGS =	-Wall @CFLAGS@ $(ARCH) $(DEBUGGER) $(OPTIMIZER) $(PROFILER)
 #LDFLAGS =	@LDFLAGS@ $(PROFILER)
 
diff -urN mad-0.14.2b.orig/libmad/Makefile.in mad-0.14.2b/libmad/Makefile.in
--- mad-0.14.2b.orig/libmad/Makefile.in	2001-11-09 08:03:06.000000000 +0100
+++ mad-0.14.2b/libmad/Makefile.in	2001-11-09 08:03:06.000000000 +0100
@@ -97,6 +97,9 @@
 libmad_la_DEPENDENCIES = @ASO_OBJS@
 libmad_la_LIBADD = @ASO_OBJS@
 
+pkgconfigdir = $(libdir)/pkgconfig
+pkgconfig_DATA = mad.pc
+
 #CFLAGS =	-Wall @CFLAGS@ $(ARCH) $(DEBUGGER) $(OPTIMIZER) $(PROFILER)
 #LDFLAGS =	@LDFLAGS@ $(PROFILER)
 
@@ -115,7 +118,7 @@
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
 mkinstalldirs = $(SHELL) $(top_srcdir)/../mkinstalldirs
 CONFIG_HEADER = config.h
-CONFIG_CLEAN_FILES = 
+CONFIG_CLEAN_FILES =  mad.pc
 LTLIBRARIES =  $(lib_LTLIBRARIES)
 
 
@@ -130,10 +133,11 @@
 LTCOMPILE = $(LIBTOOL) --mode=compile $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
 CCLD = $(CC)
 LINK = $(LIBTOOL) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(LDFLAGS) -o $@
+DATA =  $(pkgconfig_DATA)
 HEADERS =  $(include_HEADERS)
 
 DIST_COMMON =  ./stamp-h.in Makefile.am Makefile.in acconfig.h \
-acinclude.m4 aclocal.m4 config.h.in configure configure.in
+acinclude.m4 aclocal.m4 config.h.in configure configure.in mad.pc.in
 
 
 DISTFILES = $(DIST_COMMON) $(SOURCES) $(HEADERS) $(TEXINFOS) $(EXTRA_DIST)
@@ -188,6 +192,8 @@
 	-rm -f config.h
 
 maintainer-clean-hdr:
+mad.pc: $(top_builddir)/config.status mad.pc.in
+	cd $(top_builddir) && CONFIG_FILES=$@ CONFIG_HEADERS= $(SHELL) ./config.status
 
 mostlyclean-libLTLIBRARIES:
 
@@ -255,6 +261,25 @@
 libmad.la: $(libmad_la_OBJECTS) $(libmad_la_DEPENDENCIES)
 	$(LINK) -rpath $(libdir) $(libmad_la_LDFLAGS) $(libmad_la_OBJECTS) $(libmad_la_LIBADD) $(LIBS)
 
+install-pkgconfigDATA: $(pkgconfig_DATA)
+	@$(NORMAL_INSTALL)
+	$(mkinstalldirs) $(DESTDIR)$(pkgconfigdir)
+	@list='$(pkgconfig_DATA)'; for p in $$list; do \
+	  if test -f $(srcdir)/$$p; then \
+	    echo " $(INSTALL_DATA) $(srcdir)/$$p $(DESTDIR)$(pkgconfigdir)/$$p"; \
+	    $(INSTALL_DATA) $(srcdir)/$$p $(DESTDIR)$(pkgconfigdir)/$$p; \
+	  else if test -f $$p; then \
+	    echo " $(INSTALL_DATA) $$p $(DESTDIR)$(pkgconfigdir)/$$p"; \
+	    $(INSTALL_DATA) $$p $(DESTDIR)$(pkgconfigdir)/$$p; \
+	  fi; fi; \
+	done
+
+uninstall-pkgconfigDATA:
+	@$(NORMAL_UNINSTALL)
+	list='$(pkgconfig_DATA)'; for p in $$list; do \
+	  rm -f $(DESTDIR)$(pkgconfigdir)/$$p; \
+	done
+
 install-includeHEADERS: $(include_HEADERS)
 	@$(NORMAL_INSTALL)
 	$(mkinstalldirs) $(DESTDIR)$(includedir)
@@ -362,20 +387,22 @@
 install-exec-am: install-libLTLIBRARIES
 install-exec: install-exec-am
 
-install-data-am: install-includeHEADERS
+install-data-am: install-pkgconfigDATA install-includeHEADERS
 install-data: install-data-am
 
 install-am: all-am
 	@$(MAKE) $(AM_MAKEFLAGS) install-exec-am install-data-am
 install: install-am
-uninstall-am: uninstall-libLTLIBRARIES uninstall-includeHEADERS
+uninstall-am: uninstall-libLTLIBRARIES uninstall-pkgconfigDATA \
+		uninstall-includeHEADERS
 uninstall: uninstall-am
-all-am: Makefile $(LTLIBRARIES) $(HEADERS) config.h
+all-am: Makefile $(LTLIBRARIES) $(DATA) $(HEADERS) config.h
 all-redirect: all-am
 install-strip:
 	$(MAKE) $(AM_MAKEFLAGS) AM_INSTALL_PROGRAM_FLAGS=-s install
 installdirs:
-	$(mkinstalldirs)  $(DESTDIR)$(libdir) $(DESTDIR)$(includedir)
+	$(mkinstalldirs)  $(DESTDIR)$(libdir) $(DESTDIR)$(pkgconfigdir) \
+		$(DESTDIR)$(includedir)
 
 
 mostlyclean-generic:
@@ -425,9 +452,11 @@
 uninstall-libLTLIBRARIES install-libLTLIBRARIES mostlyclean-compile \
 distclean-compile clean-compile maintainer-clean-compile \
 mostlyclean-libtool distclean-libtool clean-libtool \
-maintainer-clean-libtool uninstall-includeHEADERS \
-install-includeHEADERS tags mostlyclean-tags distclean-tags clean-tags \
-maintainer-clean-tags distdir info-am info dvi-am dvi check check-am \
+maintainer-clean-libtool uninstall-pkgconfigDATA install-pkgconfigDATA \
+uninstall-includeHEADERS install-includeHEADERS tags mostlyclean-tags \
+distclean-tags clean-tags maintainer-clean-tags distdir \
+mostlyclean-depend distclean-depend clean-depend \
+maintainer-clean-depend info-am info dvi-am dvi check check-am \
 installcheck-am installcheck all-recursive-am install-exec-am \
 install-exec install-data-am install-data install-am install \
 uninstall-am uninstall all-redirect all-am all installdirs \
diff -urN mad-0.14.2b.orig/libmad/configure mad-0.14.2b/libmad/configure
--- mad-0.14.2b.orig/libmad/configure	2001-11-03 20:27:11.000000000 +0100
+++ mad-0.14.2b/libmad/configure	2001-11-03 20:27:11.000000000 +0100
@@ -2816,7 +2816,7 @@
 
 cat >> $CONFIG_STATUS <<EOF
 
-CONFIG_FILES=\${CONFIG_FILES-"Makefile"}
+CONFIG_FILES=\${CONFIG_FILES-"Makefile mad.pc"}
 EOF
 cat >> $CONFIG_STATUS <<\EOF
 for ac_file in .. $CONFIG_FILES; do if test "x$ac_file" != x..; then
diff -urN mad-0.14.2b.orig/libmad/configure.in mad-0.14.2b/libmad/configure.in
--- mad-0.14.2b.orig/libmad/configure.in	2001-11-03 20:15:07.000000000 +0100
+++ mad-0.14.2b/libmad/configure.in	2001-11-03 20:15:07.000000000 +0100
@@ -374,4 +374,4 @@
 LTLIBOBJS=`echo "$LIBOBJS" | sed -e 's/\.o/.lo/g'`
 AC_SUBST(LTLIBOBJS)
 
-AC_OUTPUT(Makefile)
+AC_OUTPUT(Makefile mad.pc)
diff -urN mad-0.14.2b.orig/libmad/mad.pc.in mad-0.14.2b/libmad/mad.pc.in
--- mad-0.14.2b.orig/libmad/mad.pc.in	1970-01-01 01:00:00.000000000 +0100
+++ mad-0.14.2b/libmad/mad.pc.in	1970-01-01 01:00:00.000000000 +0100
@@ -0,0 +1,11 @@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+libdir=@libdir@
+includedir=@includedir@
+
+Name: mad
+Description: MPEG Audio Decoder
+Requires:
+Version: 0.14.2b
+Libs: -L${libdir} -lmad -lm
+Cflags: -I${includedir}
