diff -Naur ncurses/c++/cursesw.cc ncurses-5.2/c++/cursesw.cc
--- ncurses/c++/cursesw.cc	Sat Sep  2 20:55:31 2000
+++ ncurses-5.2/c++/cursesw.cc	Tue Apr  3 11:16:59 2001
@@ -52,7 +52,8 @@
     int result = wgetstr(w, buf);
     if (result == OK) {
 	strstreambuf ss(buf, sizeof(buf));
-	result = ss.vscan(fmt, (_IO_va_list)args);
+	result = ss.vscan(fmt, args);
+//	result = ss.vscan(fmt, (_IO_va_list)args);
     }
     va_end(args);
     return result;
@@ -74,7 +75,8 @@
 	result = wgetstr(w, buf);
 	if (result == OK) {
 	    strstreambuf ss(buf, sizeof(buf));
-	    result = ss.vscan(fmt, (_IO_va_list)args);
+	    result = ss.vscan(fmt, args);
+//	    result = ss.vscan(fmt, (_IO_va_list)args);
 	}
     }
     va_end(args);
diff -Naur ncurses/ncurses/Makefile.in ncurses-5.2/ncurses/Makefile.in
--- ncurses/ncurses/Makefile.in	Sun Oct 15 00:55:35 2000
+++ ncurses-5.2/ncurses/Makefile.in	Tue Apr  3 11:18:58 2001
@@ -84,8 +84,11 @@
 
 CCFLAGS		= $(CPPFLAGS) $(CFLAGS)
 
+HOSTCPPFLAGS	= -I../ncurses -I$(srcdir) -I. -I../include \
+		  -DHAVE_CONFIG_H
+
 HOSTCC		= @BUILD_CC@
-HOSTCCFLAGS	= @CFLAGS@ $(CPPFLAGS)
+HOSTCCFLAGS	= $(HOSTCPPFLAGS)
 HOSTLDFLAGS	= @LDFLAGS@ @LIBS@
 
 CFLAGS_LIBTOOL	= $(CCFLAGS)
