diff -Naur nfs-utils-1.1.1.orig/Makefile.am nfs-utils-1.1.1/Makefile.am
--- nfs-utils-1.1.1.orig/Makefile.am	2006-07-08 02:04:32.000000000 +0200
+++ nfs-utils-1.1.1/Makefile.am	2007-11-21 21:55:05.000000000 +0200
@@ -42,17 +42,19 @@
 
 ACLOCAL_AMFLAGS = -I aclocal
 
-install-data-hook:
-	if [ ! -d $(DESTDIR)$(statedir) ]; then mkdir -p $(DESTDIR)$(statedir); fi
-	touch $(DESTDIR)$(statedir)/xtab; chmod 644 $(DESTDIR)$(statedir)/xtab
-	touch $(DESTDIR)$(statedir)/etab; chmod 644 $(DESTDIR)$(statedir)/etab
-	touch $(DESTDIR)$(statedir)/rmtab; chmod 644 $(DESTDIR)$(statedir)/rmtab
-	mkdir -p $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak
-	touch $(DESTDIR)$(statedir)/state
-	chmod go-rwx $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak $(DESTDIR)$(statedir)/state
-	-chown $(statduser) $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak $(DESTDIR)$(statedir)/state
-
-uninstall-hook:
+ install-data-hook:
+	if [ ! -d $(DESTDIR)$(statedir) ]; then mkdir -p $(DESTDIR)$(statedir); fi
+	touch $(DESTDIR)$(statedir)/xtab; chmod 644 $(DESTDIR)$(statedir)/xtab
+	if [ ! -e $(DESTDIR)$(statedir)/etab ]; then \
+		touch $(DESTDIR)$(statedir)/etab; \
+		chmod 644 $(DESTDIR)$(statedir)/etab; fi
+	touch $(DESTDIR)$(statedir)/rmtab; chmod 644 $(DESTDIR)$(statedir)/rmtab
+	mkdir -p $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak
+	touch $(DESTDIR)$(statedir)/state
+	chmod go-rwx $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak $(DESTDIR)$(statedir)/state
+#	chown $(statduser) $(DESTDIR)$(statedir)/sm $(DESTDIR)$(statedir)/sm.bak $(DESTDIR)$(statedir)/state
+  
+  uninstall-hook:
 	rm $(DESTDIR)$(statedir)/xtab
 	rm $(DESTDIR)$(statedir)/etab
 	rm $(DESTDIR)$(statedir)/rmtab
diff -Naur nfs-utils-1.1.1.orig/tools/rpcgen/Makefile.am nfs-utils-1.1.1/tools/rpcgen/Makefile.am
--- nfs-utils-1.1.1.orig/tools/rpcgen/Makefile.am	2006-07-08 02:04:32.000000000 +0200
+++ nfs-utils-1.1.1/tools/rpcgen/Makefile.am	2007-11-27 08:43:56.000000000 +0200
@@ -1,6 +1,14 @@
 ## Process this file with automake to produce Makefile.in
 
-CC=$(CC_FOR_BUILD)
+# Compile for execution on the build host, not the target. Yes, this is not
+# completely clean.
+# The warning:
+# 	tools/rpcgen/Makefile.am:6: `CFLAGS' is a user variable, you should not override it;
+# 	tools/rpcgen/Makefile.am:6: use `AM_CFLAGS' instead.
+# should be ignored.
+
+CC=gcc
+CFLAGS=
 LIBTOOL = @LIBTOOL@ --tag=CC
 
 noinst_PROGRAMS = rpcgen
--- nfs-utils-1.1.1.orig/support/nfs/xlog.c
+++ nfs-utils-1.1.1/support/nfs/xlog.c
@@ -133,9 +133,13 @@
 void
 xlog_backend(int kind, const char *fmt, va_list args)
 {
+	va_list args2;
+
 	if (!(kind & (L_ALL)) && !(logging && (kind & logmask)))
 		return;
 
+	va_copy(args2, args);
+
 	if (log_syslog) {
 		switch (kind) {
 		case L_FATAL:
@@ -172,10 +176,12 @@
 		fprintf(stderr, "%s: ", log_name);
 #endif
 
-		vfprintf(stderr, fmt, args);
+		vfprintf(stderr, fmt, args2);
 		fprintf(stderr, "\n");
 	}
 
+	va_end(args2);
+
 	if (kind == L_FATAL)
 		exit(1);
 }
