diff -Naur portmap-5-orig/Makefile portmap5-new/Makefile
--- portmap-5-orig/Makefile	2005-10-11 17:04:25.000000000 +0200
+++ portmap5-new/Makefile	2006-09-02 21:00:53.000000000 +0200
@@ -15,9 +15,9 @@
 # no access control tables. The local system, since it runs the portmap
 # daemon, is always treated as an authorized host.
 
-HOSTS_ACCESS= -DHOSTS_ACCESS
+#HOSTS_ACCESS= -DHOSTS_ACCESS
 #WRAP_LIB = $(WRAP_DIR)/libwrap.a
-WRAP_LIB = -lwrap
+#WRAP_LIB = -lwrap
 
 # Comment out if your RPC library does not allocate privileged ports for
 # requests from processes with root privilege, or the new portmap will
@@ -72,7 +72,7 @@
 # With verbose logging on, HP-UX 9.x and AIX 4.1 leave zombies behind when
 # SIGCHLD is not ignored. Enable next macro for a fix.
 #
-ZOMBIES = -DIGNORE_SIGCHLD	# AIX 4.x, HP-UX 9.x
+#ZOMBIES = -DIGNORE_SIGCHLD	# AIX 4.x, HP-UX 9.x
 
 # Uncomment the following macro if your system does not have u_long.
 #
@@ -82,7 +82,7 @@
 # libwrap.a object library. WRAP_DIR should specify the directory with
 # that library.
 
-WRAP_DIR= $(TCPD_DIR)
+#WRAP_DIR= $(TCPD_DIR)
 
 # Auxiliary object files that may be missing from your C library.
 #
@@ -143,14 +143,15 @@
 	cc $(CFLAGS) -DTEST -o $@ get_myaddress.c $(LIBS)
 
 install: all
-	install -o root -g root -m 0755 -s portmap ${BASEDIR}/sbin
-	install -o root -g root -m 0755 -s pmap_dump ${BASEDIR}/sbin
-	install -o root -g root -m 0755 -s pmap_set ${BASEDIR}/sbin
-	install -o root -g root -m 0644 portmap.8 ${BASEDIR}/usr/share/man/man8
-	install -o root -g root -m 0644 pmap_dump.8 ${BASEDIR}/usr/share/man/man8
-	install -o root -g root -m 0644 pmap_set.8 ${BASEDIR}/usr/share/man/man8
-	cat BLURB >${BASEDIR}/usr/share/doc/portmap/portmapper.txt
-	gzip -9f ${BASEDIR}/usr/share/doc/portmap/portmapper.txt
+	install -m 0755 -s portmap ${BASEDIR}/sbin
+	install -m 0755 -s pmap_dump ${BASEDIR}/sbin
+	install -m 0755 -s pmap_set ${BASEDIR}/sbin
+	install -d ${BASEDIR}/usr/share/man/man8
+	install -m 0444 portmap.8 ${BASEDIR}/usr/share/man/man8
+	install -m 0444 pmap_dump.8 ${BASEDIR}/usr/share/man/man8
+	install -m 0444 pmap_set.8 ${BASEDIR}/usr/share/man/man8
+	install -d ${BASEDIR}/usr/share/doc/portmap
+	install -m 0444 BLURB ${BASEDIR}/usr/share/doc/portmap/portmapper.txt
 
 
 lint:	
diff -Naur portmap-5-orig/pmap_check.c portmap5-new/pmap_check.c
--- portmap-5-orig/pmap_check.c	2005-10-11 17:04:25.000000000 +0200
+++ portmap5-new/pmap_check.c	2006-09-02 21:00:12.000000000 +0200
@@ -48,7 +48,7 @@
 #endif
 #include <sys/types.h>
 #include <unistd.h>
-#include <tcpd.h>
+/*#include <tcpd.h>*/
 
 extern char *inet_ntoa();
 
