diff -rup procps-3.2.7-orig/proc/escape.c procps-3.2.7/proc/escape.c
--- procps-3.2.7-orig/proc/escape.c	2005-01-05 21:50:26.000000000 +0100
+++ procps-3.2.7/proc/escape.c	2007-11-19 09:11:55.000000000 +0100
@@ -15,6 +15,9 @@
 #include "escape.h"
 #include "readproc.h"
 
+#undef __GNU_LIBRARY__
+#define __GNU_LIBRARY__ 5
+
 #if (__GNU_LIBRARY__ >= 6)
 # include <wchar.h>
 # include <wctype.h>
Files procps-3.2.7-orig/proc/escape.o and procps-3.2.7/proc/escape.o differ
Files procps-3.2.7-orig/proc/libproc-3.2.7.so and procps-3.2.7/proc/libproc-3.2.7.so differ
diff -rup procps-3.2.7-orig/ps/global.c procps-3.2.7/ps/global.c
--- procps-3.2.7-orig/ps/global.c	2005-10-30 01:43:34.000000000 +0200
+++ procps-3.2.7/ps/global.c	2007-11-19 09:14:25.000000000 +0100
@@ -28,7 +28,7 @@
 #include "../proc/version.h"
 #include "../proc/sysinfo.h"
 
-
+#undef __GNU_LIBRARY__
 #ifndef __GNU_LIBRARY__
 #define __GNU_LIBRARY__ -1
 #endif
diff -uNr procps-3.2.8.org/Makefile procps-3.2.8/Makefile
--- procps-3.2.8.org/Makefile	2009-05-10 20:50:48.000000000 +0200
+++ procps-3.2.8/Makefile	2009-11-03 14:15:19.000000000 +0100
@@ -127,7 +127,7 @@
 # produce separate executables for ppc and ppc64, s390 and s390x,
 # i386 and x86-64, mips and mips64, sparc and sparc64, and so on.
 # Failure to do so will cause data corruption.
-m64 := $(call check_gcc,-m64,$(call check_gcc,-mabi=64,))
+m64 := $(call check_gcc,,$(call check_gcc,-mabi=64,))
 ALL_CFLAGS += $(m64)
 
 ALL_CFLAGS += $(call check_gcc,-Wdeclaration-after-statement,)
