diff -aur thttpd-2.20c.orig/Makefile.in thttpd-2.20c/Makefile.in
--- thttpd-2.20c.orig/Makefile.in	Tue Jul  4 18:21:32 2000
+++ thttpd-2.20c/Makefile.in	Sat Aug  3 02:14:41 2002
@@ -30,16 +30,16 @@
 prefix = @prefix@
 exec_prefix = @exec_prefix@
 # Pathname of directory to install the binary.
-BINDIR = @sbindir@
+BINDIR = @bindir@
 # Pathname of directory to install the man page.
 MANDIR = @mandir@
 # Pathname of directory to install the CGI programs.
-WEBDIR = $(prefix)/www
+WEBDIR = $(prefix)/share/tuxbox/neutrino/httpd
 
 # CONFIGURE: The group that the web directory belongs to.  This is so that
 # the makeweb program can be installed set-group-id to that group, and make
 # subdirectories.  If you're not going to use makeweb, ignore this.
-WEBGROUP =	www
+WEBGROUP =	nogroup
 
 # CONFIGURE: Directory for CGI executables.
 CGIBINDIR =	$(WEBDIR)/cgi-bin
@@ -111,10 +111,11 @@
 install:	installthis install-man installsubdirs
 
 installthis:
-	$(INSTALL) -m 555 -o bin -g bin thttpd $(DESTDIR)$(BINDIR)
+	$(INSTALL) -m 555 thttpd $(DESTDIR)$(BINDIR)
 
 install-man:
-	$(INSTALL) -m 444 -o bin -g bin thttpd.8 $(DESTDIR)$(MANDIR)/man8
+	$(INSTALL) -d $(DESTDIR)$(MANDIR)/man8
+	$(INSTALL) -m 444 thttpd.8 $(DESTDIR)$(MANDIR)/man8
 
 installsubdirs:
 	for i in $(SUBDIRS) ; do ( \
Only in thttpd-2.20c: config.cache
diff -aur thttpd-2.20c.orig/config.h thttpd-2.20c/config.h
--- thttpd-2.20c.orig/config.h	Tue Sep 26 23:31:16 2000
+++ thttpd-2.20c/config.h	Fri Aug  2 18:57:56 2002
@@ -68,7 +68,7 @@
 /* When virtual hosting, enable the central directory on every host: */
 #define CGI_PATTERN "/*/cgi-bin/*"
 #endif
-
+#define CGI_PATTERN "/cgi-bin/*"
 /* CONFIGURE: How many seconds to allow CGI programs to run before killing
 ** them.  This is in case someone writes a CGI program that goes into an
 ** infinite loop, or does a massive database lookup that would take hours,
@@ -157,13 +157,14 @@
 ** In any case, if thttpd is started with the -h flag, that name always
 ** gets used.
 */
+#define SERVER_NAME "dbox2"
+
 #ifdef notdef
 #define SERVER_NAME "your.hostname.here"
 #define SERVER_NAME_LIST \
     "canonical.name.here/alternate.name.here", \
     "canonical.name.two/alternate.name.two"
 #endif
-
 /* CONFIGURE: Define this if you want to always chroot(), without having
 ** to give the -r command line flag.  Some people like this as a security
 ** measure, to prevent inadvertant exposure by accidentally running without -r.
@@ -210,7 +211,7 @@
 ** initializing.  If this user (or the one specified by the -u flag) does
 ** not exist, the program will refuse to run.
 */
-#define DEFAULT_USER "nobody"
+#define DEFAULT_USER "root"
 
 /* CONFIGURE: When started as root, the program can automatically chdir()
 ** to the home directory of the user specified by -u or DEFAULT_USER.
@@ -249,12 +250,12 @@
 
 /* CONFIGURE: $PATH to use for CGI programs.
 */
-#define CGI_PATH "/usr/local/bin:/usr/ucb:/bin:/usr/bin"
+#define CGI_PATH "/bin:/var:/lib"
 
 /* CONFIGURE: If defined, $LD_LIBRARY_PATH to use for CGI programs.
 */
 #ifdef notdef
-#define CGI_LD_LIBRARY_PATH "/usr/local/lib:/usr/lib"
+#define CGI_LD_LIBRARY_PATH "/bin:/sbin:/lib"
 #endif
 
 /* CONFIGURE: How often to run the occasional cleanup job.
Only in thttpd-2.20c: config.log
diff -aur thttpd-2.20c.orig/extras/Makefile.in thttpd-2.20c/extras/Makefile.in
--- thttpd-2.20c.orig/extras/Makefile.in	Wed Sep 15 18:00:14 1999
+++ thttpd-2.20c/extras/Makefile.in	Fri Aug  2 18:57:56 2002
@@ -68,8 +68,8 @@
 install:	all
 	rm -f $(BINDIR)/makeweb $(BINDIR)/htpasswd $(BINDIR)/syslogtocern
 	cp makeweb $(BINDIR)/makeweb
-	chgrp $(WEBGROUP) $(BINDIR)/makeweb
-	chmod 2755 $(BINDIR)/makeweb
+#	chgrp $(WEBGROUP) $(BINDIR)/makeweb
+#	chmod 2755 $(BINDIR)/makeweb
 	cp htpasswd $(BINDIR)/htpasswd
 	cp syslogtocern $(BINDIR)/syslogtocern
 	rm -f $(MANDIR)/man1/makeweb.1
diff -aur thttpd-2.20c.orig/extras/htpasswd.c thttpd-2.20c/extras/htpasswd.c
--- thttpd-2.20c.orig/extras/htpasswd.c	Thu May 28 23:23:48 1998
+++ thttpd-2.20c/extras/htpasswd.c	Fri Aug  2 18:57:56 2002
@@ -173,8 +173,8 @@
         exit(0);
     } else if(argc != 3) usage();
 
-    tn = tmpnam(NULL);
-    if(!(tfp = fopen(tn,"w"))) {
+//   tn = tmpnam(NULL);
+    if(!(tfp = tmpfile())) {
         fprintf(stderr,"Could not open temp file.\n");
         exit(1);
     }
diff -aur thttpd-2.20c.orig/libhttpd.c thttpd-2.20c/libhttpd.c
--- thttpd-2.20c.orig/libhttpd.c	Thu Nov 22 07:06:23 2001
+++ thttpd-2.20c/libhttpd.c	Fri Aug  2 18:57:56 2002
@@ -35,6 +35,7 @@
 #include <sys/stat.h>
 #include <sys/time.h>
 
+#include <time.h>
 #include <ctype.h>
 #include <errno.h>
 #include <fcntl.h>
diff -aur thttpd-2.20c.orig/timers.h thttpd-2.20c/timers.h
--- thttpd-2.20c.orig/timers.h	Thu Sep 14 23:10:43 2000
+++ thttpd-2.20c/timers.h	Fri Aug  2 18:57:56 2002
@@ -32,7 +32,7 @@
 
 #ifndef INFTIM
 #define INFTIM -1
-#endif INFTIM
+#endif /* INFTIM */
 
 /* ClientData is a random value that tags along with a timer.  The client
 ** can use it for whatever, and it gets passed to the callback when the
