diff -Naur tinylogin-0.80.orig/Config.h tinylogin-0.80/Config.h
--- tinylogin-0.80.orig/Config.h	Tue Dec 26 22:46:28 2000
+++ tinylogin-0.80/Config.h	Mon Dec 10 16:59:01 2001
@@ -5,13 +5,13 @@
 
 #define TLG_ADDUSER
 #define TLG_ADDGROUP
-#define TLG_DELUSER
-#define TLG_DELGROUP
+//#define TLG_DELUSER
+//#define TLG_DELGROUP
 #define TLG_LOGIN
 #define TLG_SU
 //#define TLG_SULOGIN
 #define TLG_PASSWD
-#define TLG_GETTY
+//#define TLG_GETTY
 //#define TLG_VLOCK
 //
 //
@@ -22,10 +22,10 @@
 //
 //
 // Enable using shadow passwords
-#define TLG_FEATURE_SHADOWPASSWDS
+//#define TLG_FEATURE_SHADOWPASSWDS
 //
 // Enable checking of /etc/securetty by login
-#define TLG_FEATURE_SECURETTY
+//#define TLG_FEATURE_SECURETTY
 //
 // Enable using md5 passwords
 //#define TLG_FEATURE_MD5_PASSWORDS
