diff -urN zlib-1.1.4.orig/Makefile zlib-1.1.4/Makefile
--- zlib-1.1.4.orig/Makefile	Mon Mar 11 14:58:30 2002
+++ zlib-1.1.4/Makefile	Sat Jun  1 19:41:22 2002
@@ -12,20 +12,20 @@
 # To install in $HOME instead of /usr/local, use:
 #    make install prefix=$HOME
 
-CC=cc
+CC=gcc
 
-CFLAGS=-O
+CFLAGS=-fPIC -O3 -DHAVE_UNISTD_H -DUSE_MMAP
 #CFLAGS=-O -DMAX_WBITS=14 -DMAX_MEM_LEVEL=7
 #CFLAGS=-g -DDEBUG
 #CFLAGS=-O3 -Wall -Wwrite-strings -Wpointer-arith -Wconversion \
 #           -Wstrict-prototypes -Wmissing-prototypes
 
 LDFLAGS=-L. -lz
-LDSHARED=$(CC)
-CPP=$(CC) -E
+LDSHARED=gcc -shared -Wl,-soname,libz.so.1
+CPP=gcc -E
 
 VER=1.1.4
-LIBS=libz.a
+LIBS=libz.so.1.1.4
 SHAREDLIB=libz.so
 
 AR=ar rc
@@ -33,10 +33,10 @@
 TAR=tar
 SHELL=/bin/sh
 
-prefix = /usr/local
-exec_prefix = ${prefix}
-libdir = ${exec_prefix}/lib
-includedir = ${prefix}/include
+prefix =/usr/local
+exec_prefix =${prefix}
+libdir =${exec_prefix}/lib
+includedir =${prefix}/include
 
 OBJS = adler32.o compress.o crc32.o gzio.o uncompr.o deflate.o trees.o \
        zutil.o inflate.o infblock.o inftrees.o infcodes.o infutil.o inffast.o
diff -urN zlib-1.1.4.orig/Makefile.in zlib-1.1.4/Makefile.in
--- zlib-1.1.4.orig/Makefile.in	Mon Mar 11 14:58:30 2002
+++ zlib-1.1.4/Makefile.in	Sat Jun  1 19:43:26 2002
@@ -58,7 +58,10 @@
   contrib/minizip/[CM]*[pe] contrib/minizip/*.[ch] contrib/minizip/*.[td]?? \
   contrib/delphi*/*.???
 
-all: example minigzip
+all: $(SHAREDLIB).$(VER)
+
+.c.o:
+	$(CC) -c $(CFLAGS) -fPIC -o $@ $*.c
 
 test: all
 	@LD_LIBRARY_PATH=.:$(LD_LIBRARY_PATH) ; export LD_LIBRARY_PATH; \
@@ -81,7 +84,7 @@
 	rm -f _match.s
 
 $(SHAREDLIB).$(VER): $(OBJS)
-	$(LDSHARED) -o $@ $(OBJS)
+	$(LDSHARED) -o $@ $(OBJS) -lc
 	rm -f $(SHAREDLIB) $(SHAREDLIB).1
 	ln -s $@ $(SHAREDLIB)
 	ln -s $@ $(SHAREDLIB).1
diff -urN zlib-1.1.4.orig/configure zlib-1.1.4/configure
--- zlib-1.1.4.orig/configure	Wed Jul  8 20:19:35 1998
+++ zlib-1.1.4/configure	Sat Jun  1 19:43:26 2002
@@ -72,7 +72,7 @@
   SFLAGS=${CFLAGS-"-fPIC -O3"}
   CFLAGS="$cflags"
   case `(uname -s || echo unknown) 2>/dev/null` in
-  Linux | linux) LDSHARED=${LDSHARED-"gcc -shared -Wl,-soname,libz.so.1"};;
+  Linux | linux) LDSHARED=${LDSHARED-"$CC -shared -Wl,-soname,libz.so.1"};;
   *)             LDSHARED=${LDSHARED-"gcc -shared"};;
   esac
 else
