diff -Naur zlib-1.2.1/configure zlib-1.2.1/configure.cygwin
--- zlib-1.2.1/configure	2004-01-20 14:14:20.906250000 +0100
+++ zlib-1.2.1/configure.cygwin	2004-01-20 14:14:40.625000000 +0100
@@ -77,7 +77,7 @@
   CFLAGS="$cflags"
   case `(uname -s || echo unknown) 2>/dev/null` in
   Linux | linux | GNU | GNU/*) LDSHARED=${LDSHARED-"$cc -shared -Wl,-soname,libz.so.1"};;
-  CYGWIN* | Cygwin* | cygwin* )
+  CYGWIN* | Cygwin* | cygwin* ) LDSHARED=${LDSHARED-"$cc -shared -Wl,-soname,libz.so.1"}
              EXE='.exe';;
   QNX*)  # This is for QNX6. I suppose that the QNX rule below is for QNX2,QNX4
          # (alain.bonnefoy@icbt.com)
